// citi/forms/dropdown/dropdown-option/dropdown-option.component.ts
import { Component, OnInit, Input } from '@angular/core';
import { DropdownService } from "../dropdown-service/dropdown.service";

/**
 * @Component: Dropdown Option
 * @Description: used to create an option inside a citi-dropdown component
 * from Bootstrap.
*/
@Component({
  selector: 'citi-dropdown-option',
  template: ""
})
export class DropdownOptionComponent implements OnInit {

    // Constructor 
    constructor(private dropdownService: DropdownService){}

    // Whether this option is disabled or not
    isDisabled: boolean = false;

    // Whether this option is selected or not
    isSelected: boolean = false;

    // On Init
    ngOnInit(){
        this.dropdownService.addOption(this);
    }

    /**
     * @Property: value 
    * @Description: value for the dropdown option
    * @Type: string 
    */
    @Input() value: string;

    /**
     * @Property: label 
    * @Description: label for the dropdown option
    * @Type: string 
    */
    @Input() label: string;

    /**
     * @Property: selected 
    * @Description: whether the dropdown option is selected or not
    * @Type: boolean 
    */
    @Input()
    set selected ( selected: boolean | string ){
        if ( selected == true || selected == "true" ){
            this.isSelected = true;
        }
    }

    /**
     * @Property: disabled 
    * @Description: whether the dropdown option is disabled or not
    * @Type: boolean 
    */
    @Input()
    set disabled ( disabled: boolean | string ){
        if ( disabled == true || disabled == "true" ){
            this.isDisabled = true;
        }
    }
}
