// citi/photo/photo.component.ts
import { Component, OnInit, Input } from '@angular/core';

/**
 * @Component: Photo Content
 * @Description: This is the photo component and 
 * is used display an image with a quote
*/

@Component({
  selector: 'citi-photo',
  template: require('./photo.component.html'),
  styles: [ require('./photo.component.css') ]
})

export class PhotoComponent implements OnInit {

/**
  * @Property: imgUrl 
  * @Description: This defines the URl for the image
  * @Type: string 
*/

@Input() imgUrl: string; 

/**
  * @Property: imgQuote 
  * @Description: This defines the Quote for the image
  * @Type: string 
*/

@Input() imgQuote: string; 

/**
  * @Property: imgPosition 
  * @Description: This defines the Position for the image
  * @Type: string 
*/

@Input() imgPosition: string; 
  
  private pic: any = {
    position : "left",
    url : "https://citi-ddl-ui.criticalmass.com/images/pull-out-quote-v2.png",
    quote : "F1.0 Photo Content, lorem ipsum dolor sit amet, eu mel altera mandamus. Legimus definiebas deterruisset id sea, ut nostro mel ipsum dolor" 
  };

/**
  * @Function: setUrl
  * @Description: This function is to set the URL for the image.
  * @Return: Void
*/

public setUrl(){
  console.log("URL:",this.imgUrl);
      console.log("URL:",this.pic.url);
    if(this.imgUrl){
   this.pic.url= this.imgUrl;
    }
}

/**
  * @Function: setQuote
  * @Description: This function is to set the Quote for the image.
  * @Return: Void
*/

public setQuote(){
  console.log("URL:",this.imgQuote);
      console.log("URL:",this.pic.quote);
    if(this.imgQuote){
   this.pic.quote= this.imgQuote;
    }
}

/**
  * @Function: setPosition
  * @Description: This function is to set the Position for the image.
  * @Return: Void
*/

public setPosition() {
  console.log("setposition function");
  console.log("before changing",this.pic.position);
 if(this.imgPosition == "left"){
   this.pic.position="pull-out-quote pull-quote-content-p-1";
 }
 else if(this.imgPosition == "right"){
   this.pic.position="pull-out-quote pull-quote-content-p-2";
 }
else if(this.imgPosition == "centre"){
   this.pic.position="pull-out-quote-3 pull-out-quote-p-3";
 }
console.log("after changing",this.pic.position);
  }

  constructor() { 
 }

  ngOnInit() {
    this.setPosition();
    this.setQuote();
    this.setUrl();
  }
}
