// citi/accordion/accordion-section/accordion-section.component.ts
import { Component, OnInit, Input, ElementRef, ViewChild } from '@angular/core';
import { AccordionService } from "../accordion.service";

/**
 * @Component: Accordion Section
 * @Description: This component generates an accordion section
 */
@Component({
  selector: 'citi-accordion-section',
  template: require('./accordion-section.component.html'),
  styles: [ require('./accordion-section.component.scss') ]
})
export class AccordionSectionComponent implements OnInit {

  constructor(private service:AccordionService = new AccordionService()){}

  // Whether the section is open or not
  isOpen: boolean = false;

  // Opens the section
  open: Function = function(){

    // Set the section to open
    this.isOpen = true;

    // Let the service know this section opened
    this.service.openSection();
  }

  // Closes the section
  close: Function = function(){
    
    // Set the section to closed
    this.isOpen = false;

    // Let the service know this section closed
    this.service.closeSection();
  }

  // Toggles the section
  toggle: Function = function(){
    if (this.isOpen){
      this.close();
    } else {
      this.open();
    }    
  }

  // On Init
  ngOnInit() {

    // Let the service know a new accordion section was added
    this.service.addSection(this);
  }

  /**
   * @Property: title
   * @Description: The title displayed to the right of the up/down arrow
   * @Type: String
   */
  @Input() title: string;

  /**
   * @Property: summary
   * @Description: The text displayed below the up/down arrow and the title. Always displayed.
   * @Type: String
   */
  @Input() summary: string;
}
