// ddl/forms/textarea/textarea.component.js
"use strict";
var core_1 = require('@angular/core');
var textarea_component_classes_1 = require('./textarea.component.classes');
/**
 * @Component: Text Area
 * @Section: Forms
 * @Description: This is the textarea component and is used to create reusable textarea
 * from Bootstrap.
*/
var TextareaComponent = (function () {
    function TextareaComponent() {
        this.settings = new textarea_component_classes_1.TextAreaSettings;
        this.required = true;
        this.errorMessage = "Text Area is empty";
    }
    Object.defineProperty(TextareaComponent.prototype, "size", {
        /**
           * @Property: size
           * @Description: This defines the size of the textarea component
           * @Type: string (large, medium, small, xsmall)
          */
        set: function (size) {
            this.settings.size = size;
        },
        enumerable: true,
        configurable: true
    });
    TextareaComponent.prototype.focusout = function (textAreaValue) {
        if (this.required == true && textAreaValue == '') {
            this.mandatoryError = true;
        }
        else {
            this.mandatoryError = false;
        }
    };
    TextareaComponent.prototype.ngOnInit = function () {
        switch (this.settings.size) {
            case 'large':
                this.sizeClass = textarea_component_classes_1.TextAreaInputCss.Large;
                break;
            case 'medium':
                this.sizeClass = textarea_component_classes_1.TextAreaInputCss.Medium;
                break;
            case 'small':
                this.sizeClass = textarea_component_classes_1.TextAreaInputCss.Small;
                break;
            case 'xsmall':
                this.sizeClass = textarea_component_classes_1.TextAreaInputCss.Small;
                break;
            default:
                console.log(this.settings.size);
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], TextareaComponent.prototype, "idStr", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], TextareaComponent.prototype, "labelText", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], TextareaComponent.prototype, "size", null);
    TextareaComponent = __decorate([
        core_1.Component({
            selector: 'citi-textarea',
            template: require('./textarea.component.html'),
            styles: [require('./textarea.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], TextareaComponent);
    return TextareaComponent;
}());
exports.TextareaComponent = TextareaComponent;
