// angular2-nexus-uiux/services/maps.service.js
/**
 * @module CoreServicesModule
 **/
/** MapsService **/
"use strict";
var core_1 = require('@angular/core');
/**
 * The MapsService
 * @class MapsService
 **/
var MapsService = (function () {
    function MapsService() {
        /**
         * Markers.
         */
        this.markers = [];
    }
    /**
     * Creates a new map inside of the given HTML container.
     *
     * @param el DIV element
     * @param mapOptions MapOptions object specification
     */
    MapsService.prototype.initMap = function (el, mapOptions) {
        var _this = this;
        // Instances the map.
        this.map = new google.maps.Map(el, mapOptions);
        // Adds event listener resize when the window changes size.
        window.addEventListener("resize", function () { _this.resize(); });
    };
    /**
     * Resizes the map, updating its center.
     */
    MapsService.prototype.resize = function () {
        // Saves the center.
        var latLng = this.map.getCenter();
        // Triggers resize event.
        google.maps.event.trigger(this.map, "resize");
        // Restores the center.
        this.map.setCenter(latLng);
    };
    /**
     * Sets the center map.
     *
     * @param latLng The center map
     */
    MapsService.prototype.setCenter = function (latLng) {
        if (this.map != null && latLng != null) {
            // Changes the center of the map to the given LatLng.
            this.map.panTo(latLng);
        }
    };
    /**
     * Sets zoom.
     *
     * @param zoom
     */
    MapsService.prototype.setZoom = function (zoom) {
        if (this.map != null) {
            this.map.setZoom(zoom);
        }
    };
    /**
     * Adds a marker.
     *
     * @param latLng Marker position
     * @param title Tooltip
     * @param contentString InfoWindow' content
     */
    MapsService.prototype.addMarker = function (latLng, title, contentString) {
        if (this.map != null && latLng != null) {
            // Creates the marker.
            var marker = new google.maps.Marker({
                position: latLng,
                title: title
            });
            // Adds the marker to the map.
            marker.setMap(this.map);
            // Creates the info window if required.
            if (contentString != null) {
                // Sets the max width of the info window to the width of the map element.
                var width = this.map.getDiv().clientWidth;
                var infoWindow = new google.maps.InfoWindow({
                    content: contentString,
                    maxWidth: width
                });
                // Makes the info window visible.
                marker.addListener('click', function () {
                    infoWindow.open(this.map, marker);
                });
            }
            // Pushes it to the markers array.
            this.markers.push(marker);
        }
    };
    /**
     * Deletes all markers.
     */
    MapsService.prototype.deleteMarkers = function () {
        // Removes the markers from the map.
        for (var i = 0; i < this.markers.length; i++) {
            this.markers[i].setMap(null);
        }
        // Removes references to them.
        this.markers = [];
    };
    MapsService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], MapsService);
    return MapsService;
}());
exports.MapsService = MapsService;
