// ddl/forms/errors/errors.component.js
"use strict";
var core_1 = require('@angular/core');
var errors_component_classes_1 = require('./errors.component.classes');
/**
 * @Component: Errors
 * @Section: Forms
 * @Description: This is the errors component and is used to display different types of error
 * from Bootstrap.
*/
var ErrorsComponent = (function () {
    function ErrorsComponent() {
        this.settings = new errors_component_classes_1.ErrorSettings;
    }
    Object.defineProperty(ErrorsComponent.prototype, "errorType", {
        /**
          * @Property: errorType
          * @Description: This property helps to define the type of error to be displayed
          * @Type: string (criticalAlert, alertBox, noticeAlert)
         */
        set: function (errorType) {
            this.settings.errorType = errorType;
        },
        enumerable: true,
        configurable: true
    });
    ErrorsComponent.prototype.ngOnInit = function () {
        switch (this.settings.errorType) {
            case 'criticalAlert':
                this.alertTypeClass = 'alert critical';
                break;
            case 'alertBox':
                this.alertTypeClass = 'alert-box';
                break;
            case 'noticeAlert':
                this.alertTypeClass = 'alert';
                break;
            default:
                this.alertTypeClass = 'alert-box';
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], ErrorsComponent.prototype, "errorType", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ErrorsComponent.prototype, "idStr", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ErrorsComponent.prototype, "errorMessage", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ErrorsComponent.prototype, "normalMessage", void 0);
    ErrorsComponent = __decorate([
        core_1.Component({
            selector: 'citi-errors',
            template: require('./errors.component.html'),
            styles: [require('./errors.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], ErrorsComponent);
    return ErrorsComponent;
}());
exports.ErrorsComponent = ErrorsComponent;
