// ddl/in-page-navigation/nav-group/nav-group.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Nav Group
 * @Section: Navigation
 * @Description: This is in page navigation component to create
 * from Bootstrap.
*/
var NavGroupComponent = (function () {
    function NavGroupComponent() {
        // Event emitter
        this.tabChanged = new core_1.EventEmitter();
        this.tabs = [];
    }
    NavGroupComponent.prototype.addTab = function (tab) {
        this.tabs.push(tab);
        return this.tabs.length;
    };
    NavGroupComponent.prototype.countTabs = function () {
        return this.tabs.length;
    };
    NavGroupComponent.prototype.ngOnInit = function () {
        this.tabLength = this.tabs.length;
        this.defaultTab = this.activeTab ? Number.parseInt(this.activeTab, 10) : 1;
        this.selectTab(this.tabs[this.defaultTab - 1], null);
    };
    NavGroupComponent.prototype.selectTab = function (tab, event) {
        if (event) {
            event.preventDefault();
        }
        this.tabs.forEach(function (tab) {
            tab.isExpanded = false;
        });
        tab.isExpanded = true;
        // Event emitter
        this.tabChanged.emit(tab.tabIndex);
    };
    NavGroupComponent.prototype.currentTab = function () {
        this.tabs.forEach(function (tab) {
            if (tab.isExpanded) {
                return tab.tabIndex;
            }
        });
        return -1;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], NavGroupComponent.prototype, "idStr", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], NavGroupComponent.prototype, "activeTab", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], NavGroupComponent.prototype, "tabChanged", void 0);
    NavGroupComponent = __decorate([
        core_1.Component({
            selector: 'citi-nav-group',
            template: require('./nav-group.component.html'),
            //styleUrls: ['./nav-group.component.css']
            styles: [require('./nav-group.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], NavGroupComponent);
    return NavGroupComponent;
}());
exports.NavGroupComponent = NavGroupComponent;
