// citi/forms/datepicker/ng2-datepicker/src/components/ng2-datepicker.component.ts
import { Component, ElementRef, Inject, OnInit, forwardRef, Input, Output, EventEmitter, Provider } from '@angular/core';
import { NG_VALUE_ACCESSOR, ControlValueAccessor } from '@angular/forms';
import { SlimScrollOptions } from 'ng2-slimscroll/ng2-slimscroll';
import { DatePickerOptions, DateModel, CalendarDate } from '../classes/';
import * as moment from 'moment';

const Moment: any = (<any>moment).default || moment;

export const CALENDAR_VALUE_ACCESSOR: any = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => DatePickerComponent),
  multi: true
};

@Component({
  selector: 'ng2-datepicker',
  templateUrl: 'ng2-datepicker.component.html',
  styleUrls: ['ng2-datepicker.css'],
  providers: [CALENDAR_VALUE_ACCESSOR]
})

export class DatePickerComponent implements ControlValueAccessor, OnInit {
  @Input() options: DatePickerOptions;
  @Input() inputEvents: EventEmitter<{ type: string, data: string }>;
  @Input() viewFormat: string;
  @Input() prePopulateDate: string;
  @Output() outputEvents: EventEmitter<{ type: string, data: string }>;
  

  public date: DateModel;

  private opened: boolean;
  private currentDate: moment.Moment;
  private days: CalendarDate[];
  private years: number[];
  private yearPicker: boolean;
  private scrollOptions: SlimScrollOptions;

  private minDate: moment.Moment | any;
  private maxDate: moment.Moment | any;

  private onTouchedCallback: () => void = () => { };
  private onChangeCallback: (_: any) => void = () => { };

  constructor(@Inject(ElementRef) private el: ElementRef) {
    this.opened = false;
    this.currentDate;
    this.options = this.options || {};
    this.days = [];
    this.years = [];
    this.date = new DateModel({
      day: null,
      month: null,
      year: null,
      formatted: null,
      momentObj: null
    });

    this.generateYears();

    this.outputEvents = new EventEmitter<{ type: string, data: string }>();
  }

  get value(): DateModel {
    return this.date;
  }

  set value(date: DateModel) {
    if (!date) { return; }
    this.date = date;
    this.onChangeCallback(date);
  }

  ngOnInit() {
    
    this.options.initialDate =  new Date();
    
    this.options = new DatePickerOptions(this.options);
    this.scrollOptions = {
      barBackground: '#C9C9C9',
      barWidth: '7',
      gridBackground: '#C9C9C9',
      gridWidth: '2'
    };

    if(this.prePopulateDate){
      this.options.initialDate = new Date(this.prePopulateDate);
    }

    if(this.viewFormat){
      this.options.format = this.viewFormat || this.options.format;
    }

        
    this.currentDate = Moment(this.options.initialDate);
    let date = Moment(this.currentDate);

    this.value = {
      day: date.format('DD'),
      month: date.format('MMMM'),
      year: date.format('YYYY'),
      formatted: date.format(this.options.format),
      momentObj: date
    };
  
    this.generateCalendar();

    if (this.options.minDate instanceof Date) {
      this.minDate = Moment(this.options.minDate);
    } else {
      this.minDate = null;
    }

    if (this.options.maxDate instanceof Date) {
      this.maxDate = Moment(this.options.maxDate);
    } else {
      this.maxDate = null;
    }
 
    if (typeof window !== 'undefined') {
      let body = document.querySelector('body');
      body.addEventListener('click', e => {
        if (!this.opened || !e.target) { return; };
        if (this.el.nativeElement !== e.target && !this.el.nativeElement.contains((<any>e.target))) {
          this.close();
        }
      }, false);
    }

    if (this.inputEvents) {
      this.inputEvents.subscribe((e: any) => {
        if (e.type === 'action') {
          if (e.data === 'toggle') {
            this.toggle();
          }
          if (e.data === 'close') {
            this.close();
          }
          if (e.data === 'open') {
            this.open();
          }
        }

        if (e.type === 'setDate') {
          if (!(e.data instanceof Date)) {
            throw new Error(`Input data must be an instance of Date!`);
          }
          let date: moment.Moment = Moment(e.data);

          if (!date) {
            throw new Error(`Invalid date: ${e.data}`);
          }
          this.value = {
            day: date.format('DD'),
            month: date.format('MM'),
            year: date.format('YYYY'),
            formatted: date.format(this.options.format),
            momentObj: date
          };
        }
      });
    }

  }

  generateCalendar() {
    let selectedDate = this.currentDate;
    let month = selectedDate.month();
    let year = selectedDate.year();
    let n = 1;
    let firstWeekDay = (this.options.firstWeekdaySunday) ? Moment(selectedDate).date(2).day() : Moment(selectedDate).date(1).day();
  
    if (firstWeekDay !== 1) {
      n -= (firstWeekDay + 6) % 7;
    }

    this.days = [];
    

    for (let i = n; i <= Moment(selectedDate).endOf('month').date(); i += 1) {
      let currentDate: moment.Moment = Moment(`${i}.${month + 1}.${year}`, 'DD.MM.YYYY');      
      let today: boolean = (Moment().isSame(currentDate, 'day') && Moment().isSame(currentDate, 'month')) ? true : false;
      
      let selected: boolean = (selectedDate && Moment(this.date.momentObj).isSame(currentDate, 'day')) ? true : false;
      
      let enabled: boolean = true;

      if (this.minDate !== null) {
        if (this.maxDate !== null) {
          enabled = currentDate.isBetween(this.minDate, this.maxDate, 'day', '[]') ? true : false;
        } else {
          enabled = currentDate.isBefore(this.minDate, 'day') ? false : true;
        }
      } else {
        if (this.maxDate !== null) {
          enabled = currentDate.isAfter(this.maxDate, 'day') ? false : true;
        }
      }

      let day: CalendarDate = {
        day: i > 0 ? i : null,
        month: i > 0 ? month : null,
        year: i > 0 ? year : null,
        enabled: enabled,
        today: i > 0 && today ? true : false,
        selected: i > 0 && selected ? true : false,
        momentObj: currentDate
      };

      this.days.push(day);
    }
  }

  inputChange(event:any) {

  }

  selectDate(e: MouseEvent, date: moment.Moment) {

    if (e) { e.preventDefault(); }
    setTimeout(() => {

      this.value = {
        day: date.format('DD'),
        month: date.format('MMMM'),
        year: date.format('YYYY'),
        formatted: date.format(this.options.format),
        momentObj: date
      };

      this.currentDate = Moment(date);

      this.generateCalendar();
    });

      
    if (this.options.autoApply === true || this.opened === true) {
      this.opened = false;
    }
  }

  selectYear(e: MouseEvent, year: number) {
    e.preventDefault();

    setTimeout(() => {
      let date: moment.Moment = this.currentDate.year(year);
      this.value = {
        day: date.format('DD'),
        month: date.format('MM'),
        year: date.format('YYYY'),
        formatted: date.format(this.options.format),
        momentObj: date
      };
      this.yearPicker = false;
      this.generateCalendar();
    });
  }

  generateYears() {
    let date: moment.Moment = this.options.minDate || Moment().year(Moment().year() - 40);
    let toDate: moment.Moment = this.options.maxDate || Moment().year(Moment().year() + 40);
    let years = toDate.year() - date.year();

    for (let i = 0; i < years; i++) {
      this.years.push(date.year());
      date.add(1, 'year');
    }
  }

  writeValue(date: DateModel) {
    if (!date) { return; }
    this.date = date;
  }

  registerOnChange(fn: any) {
    this.onChangeCallback = fn;
  }

  registerOnTouched(fn: any) {
    this.onTouchedCallback = fn;
  }

  prevMonth() {
    this.currentDate = this.currentDate.subtract(1, 'month');
    this.generateCalendar();
  }

  nextMonth() {    
    this.currentDate = this.currentDate.add(1, 'month');
    this.generateCalendar();
  }

  today() {
    this.currentDate = Moment();
    this.selectDate(null, this.currentDate);
  }

  toggle() {
    this.opened = !this.opened;
    if (this.opened) {
      //set this.currentdate to this.date
      this.onOpen();
    }
  }

  open() {
    this.opened = true;
    this.onOpen();
  }

  close() {
    this.opened = false;
  }

  onOpen() {
    this.yearPicker = false;
  }

  openYearPicker() {
    setTimeout(() => this.yearPicker = true);
  }

}
