// ddl/forms/address/address.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Address
 * @Section: Forms
 * @Description: This is the address component and is used to create reusable address
 * from Bootstrap.
*/
var AddressComponent = (function () {
    function AddressComponent() {
        this.isHeaderTextRequired = false;
        /**
          * @Property: isStreetHidden
          * @Description: This defines to hide the street field
          * @Type: boolean
         */
        this.isStreetHidden = false;
        /**
          * @Property: isAptHidden
          * @Description: This defines to hide the apt field
          * @Type: boolean
         */
        this.isAptHidden = false;
        /**
          * @Property: isCityHidden
          * @Description: This defines to hide the city field
          * @Type: boolean
         */
        this.isCityHidden = false;
        /**
         * @Property: isStateHidden
         * @Description: This defines to hide the state field
         * @Type: boolean
        */
        this.isStateHidden = false;
        /**
          * @Property: isZipcodeHidden
          * @Description: This defines to hide the zipcode field
          * @Type: boolean
         */
        this.isZipcodeHidden = false;
        /**
          * @Property: isHomePhoneHidden
          * @Description: This defines to hide the home phone field
          * @Type: boolean
         */
        this.isHomePhoneHidden = false;
        /**
          * @Property: isWorkPhoneHidden
          * @Description: This defines to hide the work home field
          * @Type: boolean
         */
        this.isWorkPhoneHidden = false;
    }
    AddressComponent.prototype.ngOnInit = function () {
        if (this.headerText != "") {
            this.isHeaderTextRequired = true;
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "idStr", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "headerText", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "streetLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "aptLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "cityLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "stateLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "zipCodeLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "homePhoneNoLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "workPhoneNoLabel", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], AddressComponent.prototype, "stateList", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isStreetHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isAptHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isCityHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isStateHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isZipcodeHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isHomePhoneHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AddressComponent.prototype, "isWorkPhoneHidden", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "streetValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "aptValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "cityValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "stateValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "zipcodeValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "homephoneValue", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AddressComponent.prototype, "workphoneValue", void 0);
    AddressComponent = __decorate([
        core_1.Component({
            selector: 'citi-address',
            template: require('./address.component.html'),
            styles: [require('./address.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], AddressComponent);
    return AddressComponent;
}());
exports.AddressComponent = AddressComponent;
