// ddl/photo/photo.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Photo Content
 * @Section: Modules
 * @Description: This is the photo component and
 * is used display an image with a quote
*/
var PhotoComponent = (function () {
    function PhotoComponent() {
        this.pic = {
            position: "left",
            url: "https://citi-ddl-ui.criticalmass.com/images/pull-out-quote-v2.png",
            quote: "F1.0 Photo Content, lorem ipsum dolor sit amet, eu mel altera mandamus. Legimus definiebas deterruisset id sea, ut nostro mel ipsum dolor"
        };
    }
    /**
      * @Function: setUrl
      * @Description: This function is to set the URL for the image.
      * @Return: Void
    */
    PhotoComponent.prototype.setUrl = function () {
        console.log("URL:", this.imgUrl);
        console.log("URL:", this.pic.url);
        if (this.imgUrl) {
            this.pic.url = this.imgUrl;
        }
    };
    /**
      * @Function: setQuote
      * @Description: This function is to set the Quote for the image.
      * @Return: Void
    */
    PhotoComponent.prototype.setQuote = function () {
        console.log("URL:", this.imgQuote);
        console.log("URL:", this.pic.quote);
        if (this.imgQuote) {
            this.pic.quote = this.imgQuote;
        }
    };
    /**
      * @Function: setPosition
      * @Description: This function is to set the Position for the image.
      * @Return: Void
    */
    PhotoComponent.prototype.setPosition = function () {
        console.log("setposition function");
        console.log("before changing", this.pic.position);
        if (this.imgPosition == "left") {
            this.pic.position = "pull-out-quote pull-quote-content-p-1";
        }
        else if (this.imgPosition == "right") {
            this.pic.position = "pull-out-quote pull-quote-content-p-2";
        }
        else if (this.imgPosition == "centre") {
            this.pic.position = "pull-out-quote-3 pull-out-quote-p-3";
        }
        console.log("after changing", this.pic.position);
    };
    PhotoComponent.prototype.ngOnInit = function () {
        this.setPosition();
        this.setQuote();
        this.setUrl();
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], PhotoComponent.prototype, "imgUrl", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], PhotoComponent.prototype, "imgQuote", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], PhotoComponent.prototype, "imgPosition", void 0);
    PhotoComponent = __decorate([
        core_1.Component({
            selector: 'citi-photo',
            template: require('./photo.component.html'),
            styles: [require('./photo.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], PhotoComponent);
    return PhotoComponent;
}());
exports.PhotoComponent = PhotoComponent;
