// citi/tabs/tab/tab.component.ts
import { Component, Input, OnInit, HostBinding } from '@angular/core';
import { TabGroupComponent } from '../tab-group/tab-group.component';

/**
 * @Component: Tab
 * @Description: Individual tab panel to be wrapped in a citi-tab-group
 */

@Component({
  selector: 'citi-tab',
  template: require('./tab.component.html'),
  styles: [ require('./tab.component.css') ]
})
export class TabComponent implements OnInit {

  /**
   * @Property: title
   * @Description: Text shown as the clickable tab
   * @Type: String
   */
  @Input('title') tabTitle;
  
  // Set classes on host <citi-tab> tag
  @HostBinding('class.tab-pane') classTabPane = true;
  @HostBinding('class.level1') classLevel1 = true;
  @HostBinding('class.parent') classParent = true;
  @HostBinding('class.is-expanded') isExpanded = false;

  private tabgroup: TabGroupComponent;
  private lastTab: Boolean;
  public tabIndex: number;

  constructor(private parent: TabGroupComponent) {
    this.tabgroup = parent;
    this.tabIndex = this.tabgroup.addTab(this);
  }

  ngOnInit() {
    var totalTabs: number = this.tabgroup.countTabs();
    this.lastTab = this.tabIndex === totalTabs;
    this.isExpanded = this.tabIndex === this.tabgroup.defaultTab;
  }

  selectTab(event) {
    this.tabgroup.selectTab(this, event);
  }
}
