// ddl/utility/services/masking.service.js
"use strict";
var core_1 = require('@angular/core');
var utility_classes_1 = require('../../utility/utility.classes');
var emailMask_js_1 = require('text-mask-addons/dist/emailMask.js');
var MaskingService = (function () {
    function MaskingService() {
        this.getMask = function (type) {
            switch (type) {
                case "phone":
                    return utility_classes_1.MaskTypes.phoneMask;
                case "date":
                    return utility_classes_1.MaskTypes.dateMask;
                case "zipcode":
                    return utility_classes_1.MaskTypes.zipcodeMask;
                case "dollar":
                    return utility_classes_1.MaskTypes.dollarMask;
                case "email":
                    return emailMask_js_1.default;
                case "ssn":
                    return utility_classes_1.MaskTypes.ssnMask;
                default:
                    return null;
            }
        };
    }
    MaskingService.prototype.getPhoneMask = function () {
        return utility_classes_1.MaskTypes.phoneMask;
    };
    MaskingService.prototype.getDateMask = function () {
        return utility_classes_1.MaskTypes.dateMask;
    };
    MaskingService.prototype.getZipcodeMask = function () {
        return utility_classes_1.MaskTypes.zipcodeMask;
    };
    MaskingService.prototype.getDollarMask = function () {
        return utility_classes_1.MaskTypes.dollarMask;
    };
    MaskingService.prototype.getEmailMask = function () {
        return emailMask_js_1.default;
    };
    MaskingService.prototype.getSSNPipe = function () {
        return utility_classes_1.createSSNPipe;
    };
    MaskingService = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], MaskingService);
    return MaskingService;
}());
exports.MaskingService = MaskingService;
