// ddl/forms/table/table.component.props.js
"use strict";
var TableProperties = (function () {
    function TableProperties() {
        this.id = '';
        this.caption = '';
        this.hideCaption = false;
        this.summary = '';
        this.tableTooltip = false;
        this.isSortable = false;
        this.isAutoInitTableSorter = false;
        this.beforeTableSorterInitialization = false;
        this.sorterConfig = '';
        this.filterContent = '';
        this.filterAligned = false;
        this.viewMore = '';
        this.isExpandable = false;
        this.isExpanded = false;
        this.hasMasterDetailsRows = false;
        this.noTopMargin = false;
        this.isFixedLayout = true;
        this.ctaText = '';
        this.ctaTooltipTitle = '';
        this.ctaTooltipBody = '';
        this.isExpandableRow = false;
        this.expandDetailsHeader = false;
        this.collapseDetailsHeader = false;
        this.tableType = false;
        this.isFixedColumnTable = false;
        this.bodyText = '';
        this.data = [];
        this.displayedHeaders = [];
        this.isInfiniteScrollEnabled = false;
        this.loadButtonText = 'Load %d more records';
        this.defaultNumberOfVisibleRecords = 10;
        this.defaultAdditionalRecords = 10;
        this.currentNumberOfVisibleRecords = 0;
        this.loadingText = 'Just a moment please.';
        this.isLoadingMoreData = false;
        this.visibleRows = [];
        this.canLoadMoreData = false;
        this.expandButtonTextClose = 'More Details';
        this.expandButtonTextOpen = 'Less Details';
        this.itemDetailCollapses = [];
        this.isZebraTable = false;
    }
    return TableProperties;
}());
exports.TableProperties = TableProperties;
