// ddl/forms/table/table.component.example.js
"use strict";
var core_1 = require('@angular/core');
var TableExampleComponent = (function () {
    function TableExampleComponent() {
        this.headerConfig = [
            { text: 'Extension', sortable: true, column: '0', isRightAligned: false },
            { text: 'XYZ', sortable: true, column: '1', isRightAligned: true },
            { text: 'ABC', sortable: false, column: '2', isRightAligned: true }];
        this.rows = [
            {
                0: '<div>Data 1</div>',
                1: 100.23,
                2: 1433588216000,
                details: 1
            },
            {
                0: '<div>Data 2</div>',
                1: 0.875623,
                2: 1432387616000,
                details: 1
            },
            [
                '<div>Data 3</div>',
                .010123,
                1461820116000
            ],
            [
                '<div>Data 4</div>',
                1873.02301,
                1423128616000
            ],
            [
                '<div>Data 5</div>',
                -93,
                1439220116000
            ],
            [
                '<div>Data 6</div>',
                100.23,
                1433588216000
            ],
            [
                '<div>Data 7</div>',
                0.875623,
                1432387616000
            ],
            [
                '<div>Data 8</div>',
                .010123,
                1461820116000
            ],
            [
                '<div>Data 9</div>',
                1873.02301,
                1423128616000
            ],
            [
                '<div>Data 10</div>',
                -93,
                1439220116000
            ],
            [
                '<div>Data 11</div>',
                100.23,
                1433588216000
            ],
            [
                '<div>Data 12</div>',
                0.875623,
                1432387616000
            ],
            [
                '<div>Data 13</div>',
                .010123,
                1461820116000
            ],
            [
                '<div>Data 14</div>',
                1873.02301,
                1423128616000
            ],
            [
                '<div>Data 15</div>',
                -93,
                1439220116000
            ],
            [
                '<div>Data 16</div>',
                100.23,
                1433588216000
            ],
            [
                '<div>Data 17</div>',
                0.875623,
                1432387616000
            ],
            [
                '<div>Data 18</div>',
                .010123,
                1461820116000
            ],
            [
                '<div>Data 19</div>',
                1873.02301,
                1423128616000
            ],
            [
                '<div>Data 20</div>',
                -93,
                1439220116000
            ]
        ];
        this.x = true;
        this.descending = false;
        this.column = '1';
        this.onEdit = new core_1.EventEmitter();
    }
    Object.defineProperty(TableExampleComponent.prototype, "sortedColumn", {
        get: function () {
            return this.column;
        },
        set: function (value) {
            this.column = value;
        },
        enumerable: true,
        configurable: true
    });
    TableExampleComponent.prototype.ngOnInit = function () {
        var _this = this;
        setTimeout(function () {
            _this.x = false;
        }, 5000);
    };
    TableExampleComponent.prototype.triggerEdit = function (index) {
        console.log(index);
        this.onEdit.emit(index);
    };
    TableExampleComponent.prototype.convertSorting = function () {
        return this.descending ? '-' + this.column : this.column;
    };
    TableExampleComponent.prototype.changeLoadingButton = function () {
        console.log(this);
    };
    TableExampleComponent.prototype.changeSorting = function (columnName) {
        if (this.column === columnName) {
            this.descending = !this.descending;
        }
        else {
            this.column = columnName;
            this.descending = false;
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], TableExampleComponent.prototype, "sortedColumn", null);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], TableExampleComponent.prototype, "onEdit", void 0);
    TableExampleComponent = __decorate([
        core_1.Component({
            selector: 'app-table-example',
            template: require('./table-example.component.html'),
            styleUrls: ['./table-example.component.css']
        }), 
        __metadata('design:paramtypes', [])
    ], TableExampleComponent);
    return TableExampleComponent;
}());
exports.TableExampleComponent = TableExampleComponent;
