// citi/forms/errors/errors.component.ts
import { Component, OnInit, Input } from '@angular/core';
import { ErrorSettings } from './errors.component.classes';

/**
 * @Component: Errors
 * @Description: This is the errors component and is used to display different types of error
 * from Bootstrap.
*/
@Component({
  selector: 'citi-errors',
  template: require('./errors.component.html'),
  styles: [ require('./errors.component.css') ]
})
export class ErrorsComponent implements OnInit {
  settings:ErrorSettings = new ErrorSettings;
  private alertTypeClass:string;
  /**
    * @Property: errorType 
    * @Description: This property helps to define the type of error to be displayed
    * @Type: string (criticalAlert, alertBox, noticeAlert)
   */
  @Input()
  set errorType ( errorType: any ){
    this.settings.errorType = errorType;
  }
  /**
    * @Property: idStr 
    * @Description: This property helps to define the id of the error section
    * @Type: string
   */
  @Input() idStr:string;
  /**
    * @Property: errorMessage 
    * @Description: This property helps to define the errorMessage of the error section
    * @Type: string
   */
  @Input() errorMessage:string;
  /**
    * @Property: normalMessage 
    * @Description: This property helps to define the normalMessage (displays with normal text) of the error section other than error message
    * @Type: string
   */
  @Input() normalMessage:string;
 
  constructor() { }

  ngOnInit() {

    switch(this.settings.errorType) {
      case 'criticalAlert':
        this.alertTypeClass = 'alert critical';
        break;
      case 'alertBox':
        this.alertTypeClass = 'alert-box';
        break;
      case 'noticeAlert':
        this.alertTypeClass = 'alert';
        break;      
      default:
      this.alertTypeClass = 'alert-box';        
    }
  }

}
