// ddl/forms/password-strength-control/password-strength-control.component.Settings.js
"use strict";
var PasswordStrengthControlSettings = (function () {
    function PasswordStrengthControlSettings() {
        this.id = '';
        this.labelText = '';
        this.errorText = '';
        this.validStatusText = '';
        this.invalidStatusText = '';
        this.initialStatusText = '';
        this.rulesHeaderText = '';
        this.userIDFieldID = '';
        this.userIDCheckMessage = '';
        this.spaceCheckMessage = '';
        this.letterCheckMessage = '';
        this.numberCheckMessage = '';
        this.lengthCheckMessage = '';
        this.minLength = '6';
        this.maxLength = '60';
        this.placeholder = '';
        this.isHidden = false;
        this.inputSize = 'large';
        this.offsetLG = '';
        this.offsetMD = '';
        this.offsetSM = '';
        this.offsetXS = '';
        this.hasHelpIcon = 'false';
    }
    return PasswordStrengthControlSettings;
}());
exports.PasswordStrengthControlSettings = PasswordStrengthControlSettings;
var PopoverSettings = (function () {
    function PopoverSettings() {
        this.popupOpacity = 0;
        this.popoverPosition = 'auto';
    }
    return PopoverSettings;
}());
exports.PopoverSettings = PopoverSettings;
