// citi/forms/textarea/textarea.component.ts
import { Component, OnInit, Input } from '@angular/core';
import { TextAreaSettings, TextAreaInputCss } from './textarea.component.classes';

/**
 * @Component: Text Area
 * @Description: This is the textarea component and is used to create reusable textarea
 * from Bootstrap.
*/
@Component({
  selector: 'citi-textarea',
  template: require('./textarea.component.html'),
  styles: [require('./textarea.component.css')]
})
export class TextareaComponent implements OnInit {
  settings: TextAreaSettings = new TextAreaSettings;
  private sizeClass: string;
  private required: boolean = true;
  private errorMessage: string = "Text Area is empty";
  private mandatoryError: boolean;
  /**
    * @Property: idStr 
    * @Description: This defines the id of the textarea component
    * @Type: string 
   */
  @Input() idStr: string;
   /**
     * @Property: labelText 
     * @Description: This defines the Label of the textarea component
     * @Type: string 
    */
  @Input() labelText: string;
  /**
     * @Property: size 
     * @Description: This defines the size of the textarea component
     * @Type: string (large, medium, small, xsmall)
    */
  @Input()
  set size(size: any) {
    this.settings.size = size;
  }
  focusout(textAreaValue) {
    if (this.required == true && textAreaValue == '') {
      this.mandatoryError = true;
    }
    else {
      this.mandatoryError = false;
    }
  }
  constructor() { }

  ngOnInit() {
    switch (this.settings.size) {
      case 'large':
        this.sizeClass = TextAreaInputCss.Large;
        break;
      case 'medium':
        this.sizeClass = TextAreaInputCss.Medium;
        break;
      case 'small':
        this.sizeClass = TextAreaInputCss.Small;
        break;
      case 'xsmall':
        this.sizeClass = TextAreaInputCss.Small;
        break;
      default:
        console.log(this.settings.size);
    }
  }
}
