// citi/citi.module.ts
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

// Call To Action (Buttons and Links)
import { CTAComponent } from "./cta/cta.component";

// Branding
import { BrandingService } from "./branding/services/branding.service";
import { FooterComponent } from "./branding/footer/footer.component";
import { FooterButlerBarComponent } from './branding/footer/butler/butler.component';
import { FooterNavigationComponent } from "./branding/footer/navigation/navigation.component";
import { SubNavigationComponent } from "./branding/footer/sub-navigation/sub-navigation.component";
import { DisclaimerBrandingComponent } from "./branding/footer/disclaimer/disclaimer.component";
import { SubFooterComponent } from "./branding/footer/sub-footer/sub-footer.component";
import { HeaderComponent } from "./branding/header/header.component";
import { ButlerBarComponent } from './branding/header/butler/butler.component';
import { BannerComponent } from "./branding/header/banner/banner.component";
import { NavigationComponent } from "./branding/header/navigation/navigation.component";
import { WelcomeBarComponent } from "./branding/header/welcome/welcome.component";

// Modal
import { ModalComponent } from "./modal/modal.component";

// Text Header (H1, H2, etc)
import { TextHeaderComponent } from "./text-header/text-header.component";

// Radios and Checkboxes
import { InputGroupModule } from "./forms/input-group/input-group.module";

// Form group
import { FormGroupComponent } from "./forms/group/group.component";

 // Accordion
 import { AccordionModule } from "./accordion/accordion.module";

 // Terms and Conditions
 import { TermsAndConditionsComponent } from "./terms/terms.component";

 // List Component
 import { 
    OrderedListComponent, 
    UnorderedListComponent, 
    CheckmarkListComponent, 
    XmarkListComponent, 
    ListItemComponent
} from "./lists/lists"

// Copy Components
import {
  CopyComponent,
  ListComponent,
  SubContentComponent,
  SubContentSectionComponent,
  ImportantCopyComponent,
  TwoColumnListComponent,
  TwoColumnListSectionComponent,
  TwoColumnContentComponent,
  TwoColumnTitleSectionComponent,
  TwoColumnTitleGridComponent,
  TitleGridSectionComponent
 } from "./copy/copy";

// Date Iput
import { DateInputComponent } from "./forms/date-input/date-input.component";

// The Rest
import { FormsModule } from '@angular/forms';
import { TableComponent } from './forms/table/table.component';
import { InputComponent } from './forms/input/input.component';
import { RowComponent } from './grid/row/row.component';
import { ColumnComponent } from './grid/column/column.component';
import { AlertComponent } from './alert/alert.component';
import { SliderComponent } from './forms/slider/slider.component';
import { ImageComponent } from './image/image.component';
import { TooltipComponent } from './tooltip/tooltip.component';
import { ErrorsComponent } from './forms/errors/errors.component';
import { IconComponent } from './icons/icon/icon.component';
import { TextareaComponent } from './forms/textarea/textarea.component';
import { PasswordStrengthControlComponent } from './forms/password-strength-control/password-strength-control.component';
import { TabGroupComponent } from './tabs/tab-group/tab-group.component';
import { TabComponent } from './tabs/tab/tab.component';
import { DropdownComponent, DropdownOptionComponent, DropdownService } from './forms/dropdown/dropdown';
import { CardInfoComponent } from './card-info/card-info.component';
import { AddressComponent } from './forms/address/address.component';
import { DisclaimerComponent } from './forms/disclaimer/disclaimer.component';
import { MessageAreaComponent } from './forms/message-area/message-area.component';
import { BottomNavigationComponent } from './forms/bottom-navigation/bottom-navigation.component';
import { JampComponent } from './jamp/jamp.component';
import { DobControlComponent } from './forms/dob-control/dob-control.component';
import { CarouselControlComponent } from './carousel-control/carousel-control.component';
import { Carousel } from './carousel-control/carousel.component';
import { CarouselSlide } from './carousel-control/carousel-slide.component';
import { FormContainerComponent } from './forms/form-container/form-container.component';
import { TextMaskModule } from 'angular2-text-mask';
import { ReactiveFormsModule } from "@angular/forms";
import { ProgressComponent } from './progress/progress.component';
import { HeroComponent } from './hero/hero.component';
import { PercentageProgressComponent } from './progress/percentageProgress.component';
import { LeftnavComponent } from './navigations/left-navigation/left-navigation-left-nav/left-nav.component';
import { ParentComponent } from './navigations/left-navigation/left-navigation-parent/left-nav-parent.component';
import { ChildComponent } from './navigations/left-navigation/left-navigation-child/left-nav-child.component';
import { SubchildComponent } from './navigations/left-navigation/left-navigation-subchild/left-nav-subchild.component';
//import { DatepickerComponent } from './forms/datepicker/datepicker.component';
//import { DatePickerComponent } from './forms/datepicker/ng2-datepicker/src/components/ng2-datepicker.component';
import { NavGroupComponent } from './in-page-navigation/nav-group/nav-group.component';
import { NavPanelComponent } from './in-page-navigation/nav-panel/nav-panel.component';
import { LeftNavLayoutComponent } from './layouts/leftnav/leftnav-layout.component';
import { SimpleLayoutComponent } from './layouts/simple/simple-layout.component';
import { CardModule } from './card-module/card-module.module';
import { ComparisonTableComponent } from './comparison-table/comparison-table.component';
import { ActionLinkComponent } from './action-link/action-link.component';
import { ActionLinkStackComponent } from './action-link-stack/action-link-stack.component';
import { ActionPrimaryMultipleRowComponent } from './action-primary-multiple-row/action-primary-multiple-row.component';
import { ActionSecondaryMultipleRowComponent } from './action-secondary-multiple-row/action-secondary-multiple-row.component';
import { ActionSecondarySingleRowComponent } from './action-secondary-single-row/action-secondary-single-row.component';
import { ActionContactUsComponent } from './action-contact-us/action-contact-us.component';
import { InputMaskDirective } from "./utility/text-mask/text-mask.directive";
import { PhotoComponent } from './photo/photo.component';

import 'rxjs';

const SERVICES: any[] = [
  BrandingService,
  DropdownService
]

const MODULES: any[] = [
  CommonModule,
  FormsModule,
  InputGroupModule,
  //TextMaskModule,
  ReactiveFormsModule,
  AccordionModule,
  CardModule
];

const COMPONENTS: any[] = [
  InputMaskDirective,
  CTAComponent,
  TextHeaderComponent,
  ModalComponent,
  FormGroupComponent,
  OrderedListComponent,
  UnorderedListComponent,
  CheckmarkListComponent,
  XmarkListComponent,
  ListItemComponent,
  CopyComponent,
  ListComponent,
  SubContentComponent,
  SubContentSectionComponent,
  ImportantCopyComponent,
  TwoColumnListComponent,
  TwoColumnListSectionComponent,
  TwoColumnContentComponent,
  TwoColumnTitleSectionComponent,
  TwoColumnTitleGridComponent,
  TitleGridSectionComponent,
  TermsAndConditionsComponent,
  TableComponent,
  InputComponent,
  RowComponent,
  ColumnComponent,
  AlertComponent,
  SliderComponent,
  ImageComponent,
  TooltipComponent,
  TooltipComponent,
  ErrorsComponent,
  IconComponent,
  TextareaComponent,
  TabGroupComponent,
  TabComponent,
  PasswordStrengthControlComponent,
  DropdownComponent,
  DropdownOptionComponent,
  CardInfoComponent,
  AddressComponent,
  DisclaimerComponent,
  MessageAreaComponent,
  BottomNavigationComponent,
  JampComponent,
  DobControlComponent,
  CarouselControlComponent,
  Carousel,
  CarouselSlide,
  FormContainerComponent,
  ProgressComponent,
  HeroComponent,
  PercentageProgressComponent,
  LeftnavComponent,
  ParentComponent,
  ChildComponent,
  SubchildComponent,
  //DatepickerComponent,
  //DateInputComponent,
  //DatePickerComponent,
  NavGroupComponent,
  NavPanelComponent,
  LeftNavLayoutComponent,
  SimpleLayoutComponent,
  FooterComponent,
  FooterButlerBarComponent,
  FooterNavigationComponent,
  SubNavigationComponent,
  DisclaimerBrandingComponent,
  SubFooterComponent,
  HeaderComponent,
  ButlerBarComponent,
  BannerComponent,
  NavigationComponent,
  WelcomeBarComponent,
  ComparisonTableComponent,
  ActionLinkComponent,
  ActionLinkStackComponent,
  ActionPrimaryMultipleRowComponent,
  ActionSecondaryMultipleRowComponent,
  ActionSecondarySingleRowComponent,
  ActionContactUsComponent,
  PhotoComponent
];

export class BrandingModule {}

@NgModule({
  imports: [
    ...MODULES
  ],
  declarations: [
    ...COMPONENTS
  ],
  exports: [
    ...COMPONENTS,
    ...MODULES
  ],
  providers: [
    ...SERVICES
  ]
})
export class CitiModule { }
