// ddl/carousel-control/carousel-slide.component.js
"use strict";
var core_1 = require('@angular/core');
var carousel_component_1 = require('./carousel.component');
var CarouselSlide = (function () {
    function CarouselSlide(carousel) {
        this.carousel = carousel;
        this.addClass = true;
    }
    CarouselSlide.prototype.ngOnInit = function () {
        this.carousel.addSlide(this);
    };
    CarouselSlide.prototype.ngOnDestroy = function () {
        //this.carousel.removeSlide(this);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], CarouselSlide.prototype, "index", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], CarouselSlide.prototype, "direction", void 0);
    __decorate([
        core_1.HostBinding('class.active'),
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], CarouselSlide.prototype, "active", void 0);
    __decorate([
        core_1.HostBinding('class.item'),
        core_1.HostBinding('class.carousel-item'), 
        __metadata('design:type', Boolean)
    ], CarouselSlide.prototype, "addClass", void 0);
    CarouselSlide = __decorate([
        core_1.Component({
            selector: 'slide',
            template: "\n    <div [class.active]=\"active\" class=\"item text-center\">\n      <ng-content></ng-content>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [carousel_component_1.Carousel])
    ], CarouselSlide);
    return CarouselSlide;
}());
exports.CarouselSlide = CarouselSlide;
