// ddl/card-info/card-info.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Card Info
 * @Section: Modules
 * @Description: The card info module.
 */
var CardInfoComponent = (function () {
    function CardInfoComponent() {
        /**
          * @Event: cardChanged
          * @Description: This event that is fired when a card is selected; returns the index of the card (first card is 0)
          * @Return: Number
          */
        this.cardChanged = new core_1.EventEmitter();
    }
    CardInfoComponent.prototype.ngOnInit = function () {
        this.startCardIndex = this.selectedIndex ? Number.parseInt(this.selectedIndex, 10) : 0;
        this.selectCard(this.startCardIndex);
    };
    CardInfoComponent.prototype.selectCard = function (cardIndex) {
        this.currentCard = this.cardList[cardIndex];
        // Extract iteratable list of details: [{label:"", value:""}]
        this.currentCardDetails = [];
        for (var detailKey in this.cardList[cardIndex].details) {
            this.currentCardDetails.push(this.cardList[cardIndex].details[detailKey]);
        }
        // Event emitter
        this.cardChanged.emit(this.currentCard);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], CardInfoComponent.prototype, "id", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Array)
    ], CardInfoComponent.prototype, "cardList", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], CardInfoComponent.prototype, "selectedIndex", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], CardInfoComponent.prototype, "readonly", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], CardInfoComponent.prototype, "showDetails", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], CardInfoComponent.prototype, "selectCardLabel", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], CardInfoComponent.prototype, "cardChanged", void 0);
    CardInfoComponent = __decorate([
        core_1.Component({
            selector: 'citi-card-info',
            template: require('./card-info.component.html'),
            styles: [require('./card-info.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], CardInfoComponent);
    return CardInfoComponent;
}());
exports.CardInfoComponent = CardInfoComponent;
