// ddl/accordion/accordion-section/accordion-section.component.js
"use strict";
var core_1 = require('@angular/core');
var accordion_service_1 = require("../accordion.service");
/**
 * @Component: Accordion Section
 * @Description: This component generates an accordion section
 */
var AccordionSectionComponent = (function () {
    function AccordionSectionComponent(service) {
        if (service === void 0) { service = new accordion_service_1.AccordionService(); }
        this.service = service;
        // Whether the section is open or not
        this.isOpen = false;
        // Opens the section
        this.open = function () {
            // Set the section to open
            this.isOpen = true;
            // Let the service know this section opened
            this.service.openSection();
        };
        // Closes the section
        this.close = function () {
            // Set the section to closed
            this.isOpen = false;
            // Let the service know this section closed
            this.service.closeSection();
        };
        // Toggles the section
        this.toggle = function () {
            if (this.isOpen) {
                this.close();
            }
            else {
                this.open();
            }
        };
    }
    // On Init
    AccordionSectionComponent.prototype.ngOnInit = function () {
        // Let the service know a new accordion section was added
        this.service.addSection(this);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AccordionSectionComponent.prototype, "title", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AccordionSectionComponent.prototype, "summary", void 0);
    AccordionSectionComponent = __decorate([
        core_1.Component({
            selector: 'citi-accordion-section',
            template: require('./accordion-section.component.html'),
            styles: [require('./accordion-section.component.scss')]
        }), 
        __metadata('design:paramtypes', [accordion_service_1.AccordionService])
    ], AccordionSectionComponent);
    return AccordionSectionComponent;
}());
exports.AccordionSectionComponent = AccordionSectionComponent;
