// citi/carousel-control/carousel.component.ts
import { Component, OnInit, Input, OnDestroy } from '@angular/core';
import { CarouselSlide } from './carousel-slide.component';
export enum Direction {UNKNOWN, NEXT, PREV}

@Component({
    selector: 'carousel',    
    template: `
    <div (mouseenter)="pause()" (mouseleave)="play()" class="carousel slide">
      <ol class="carousel-indicators" [hidden]="slides.length <= 1">
         <li *ngFor="let slideItem of slides" [class.active]="slideItem.active === true" (click)="select(slideItem)"></li>
      </ol>
      <div class="carousel-inner"><ng-content></ng-content></div>
                  <a class="left carousel-control" (click)="prev()" [hidden]="!slides.length">
                  <span class="glyphicon glyphicon-chevron-left"></span>
                  </a>
                  <a class="right carousel-control" (click)="next()" [hidden]="!slides.length">
                  <span class="glyphicon glyphicon-chevron-right"></span>
                 </a>
    </div>
  `
})
export class Carousel implements OnDestroy {
    @Input() public noWrap:boolean;
    @Input() public noPause:boolean;
    @Input() public noTransition:boolean;
    @Input() public get interval():number {
        return this._interval;
    }
    /* To set the time interval to slide the carousel image. */
    public set interval(value:number) {       
        this._interval = value;
        this.restartTimer();
    }

    private slides:Array<CarouselSlide> = [];
    private currentInterval:any;
    private isPlaying:boolean;
    private destroyed:boolean = false;
    private currentSlide:CarouselSlide;
    private _interval:number;

    public ngOnDestroy() {
        this.destroyed = true;
    }
    /*this function will be triggered when next/prev will be clicked.
     * CarouselSlide - the next slide to be displayed
     * Direction - the direction will be based on the next/prev
     * */
    public select(nextSlide:CarouselSlide, direction:Direction = Direction.UNKNOWN) {
        let nextIndex = nextSlide.index;
        if (direction === Direction.UNKNOWN) {
            direction = nextIndex > this.getCurrentIndex() ? Direction.NEXT : Direction.PREV;
        }

        // Prevent this user-triggered transition from occurring if there is already one in progress
        if (nextSlide && nextSlide !== this.currentSlide) {
            this.goNext(nextSlide, direction);
        }
    }

    private goNext(slide:CarouselSlide, direction:Direction) {
        if (this.destroyed) {
            return;
        }

        slide.direction = direction;
        slide.active = true;

        if (this.currentSlide) {
            this.currentSlide.direction = direction;
            this.currentSlide.active = false;
        }

        this.currentSlide = slide;

        // every time you change slides, reset the timer
        this.restartTimer();
    }

    private getSlideByIndex(index:number) {
        let len = this.slides.length;
        for (let i = 0; i < len; ++i) {
            if (this.slides[i].index === index) {
                return this.slides[i];
            }
        }
    }

    private getCurrentIndex() {
        return !this.currentSlide ? 0 : this.currentSlide.index;
    }

    private next() {
        let newIndex = (this.getCurrentIndex() + 1) % this.slides.length;

        if (newIndex === 0 && this.noWrap) {
            this.pause();
            return;
        }

        return this.select(this.getSlideByIndex(newIndex), Direction.NEXT);
    }

    private prev() {
        let newIndex = this.getCurrentIndex() - 1 < 0 ? this.slides.length - 1 : this.getCurrentIndex() - 1;

        if (this.noWrap && newIndex === this.slides.length - 1) {
            this.pause();
            return;
        }

        return this.select(this.getSlideByIndex(newIndex), Direction.PREV);
    }
    /* this function will be triggered to reset and initiate the interval */
    private restartTimer() {
        this.resetTimer();
        let interval = +this.interval;
        if (!isNaN(interval) && interval > 0) {
            this.currentInterval = setInterval(() => {
                let nInterval = +this.interval;
                if (this.isPlaying && !isNaN(this.interval) && nInterval > 0 && this.slides.length) {
                    this.next();
                } else {
                    this.pause();
                }
            }, interval);
        }
    }
    /* this function will be triggered to reset the interval */
    private resetTimer() {
        if (this.currentInterval) {
            console.log("this.currentInterval",this.currentInterval);
            clearInterval(this.currentInterval);
            this.currentInterval = null;
        }
    }
    /* this function will be triggered when mouseleave */
    public play() {
        if (!this.isPlaying) {
            this.isPlaying = true;
            this.restartTimer();
        }
    }
    /* this function will be triggered when mouseenter */
    public pause() {
        if (!this.noPause) {
            this.isPlaying = false;
            this.resetTimer();
        }
    }
    /* addSlide will be triggered when each carousel image is getting added in slide component */
    public addSlide(slide:CarouselSlide) {      
        slide.index = this.slides.length;
        this.slides.push(slide);
        if (this.slides.length === 1 || slide.active) {
            this.select(this.slides[this.slides.length - 1]);
            if (this.slides.length === 1) {
                this.play();
            }
        } else {
            slide.active = false;
        }
    }

    
}
