// citi/branding/footer/disclaimer/disclaimer.component.ts
import { Component, Input } from '@angular/core';

/**
 * @Component: Footer Disclaimer
 * @Description: This component generates the disclaimer section in the footer.
 */
@Component({
    selector: 'citi-footer-disclaimer',
    template: require('./disclaimer.component.html'),
    styles: [ require("./disclaimer.component.scss") ]
})

export class DisclaimerBrandingComponent {

    // This variable will cover the actual DOM logic
    config: any;

    // This function takes the BrandingConfig object and translates it to config DOM
    // This step is not necessary, but keeps the DOM markup semantic and un-bloated
    // and keeps most of the logic in the controller instead of the view
    setDisclaimerConfig = function( _config: any ){

        // Set up config object to temporarily hold our calculated values
        var disclaimerConfig: any = {};

        // If there is a valid config and a footer object is defined
        if ( _config && _config.footer ){

            // If the butler bar should be shown
            if ( _config.footer.disclaimer 
                 && _config.footer.disclaimer.show
            ){

                // If there is a footer disclaimer sections array and it has at least one section
                if ( _config.footer.disclaimer.text ){

                    // Set the disclaimer to show
                    disclaimerConfig.showDisclaimer = true;
                    
                    // Provide the sections to the config object
                    disclaimerConfig.text = _config.footer.disclaimer.text;
                }
            }

            // Set the config for use by the dom (DOM will not show until this is set);
            this.config = disclaimerConfig;
        }
    }

    /**
     * @Property: brandingConfig
     * @Description: This is a reference to the branding config JSON object retrieved from the branding API.
     * @Type: String
     */
    @Input()
    set brandingConfig( _config: any ){
        if ( _config ){
            this.setDisclaimerConfig( _config );
        }
    }
}
