// duracell/statements-routing.module.js
"use strict";
var core_1 = require('@angular/core');
var router_1 = require('@angular/router');
var statements_module_1 = require('./statements.module');
var input_component_1 = require('./input.component');
var recap_component_1 = require('./recap.component');
var confirmation_component_1 = require('./confirmation.component');
/*
    ANGULAR2-NEXUS-STARTER NOTE:
    For Routing Modules related to modules with no sub modules follow the current example
*/
exports.STATEMENTS_ROUTES = [
    /*
        ANGULAR2-NEXUS-STARTER NOTE:
        Redirect to the first page for the module.
    */
    {
        path: '',
        redirectTo: 'input',
        pathMatch: 'full'
    },
    /*
        ANGULAR2-NEXUS-STARTER NOTE:
        Specify the path and the component for every page.
    */
    {
        path: 'input',
        component: input_component_1.InputComponent
    },
    {
        path: 'recap',
        component: recap_component_1.RecapComponent
    },
    {
        path: 'confirmation',
        component: confirmation_component_1.ConfirmationComponent
    }
];
var StatementsRoutingModule = (function () {
    function StatementsRoutingModule() {
    }
    StatementsRoutingModule = __decorate([
        core_1.NgModule({
            imports: [
                router_1.RouterModule.forChild(exports.STATEMENTS_ROUTES),
                statements_module_1.StatementsModule
            ]
        }), 
        __metadata('design:paramtypes', [])
    ], StatementsRoutingModule);
    return StatementsRoutingModule;
}());
exports.StatementsRoutingModule = StatementsRoutingModule;
