// citi/in-page-navigation/nav-group/nav-group.component.ts
import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { NavPanelComponent } from '../nav-panel/nav-panel.component';

/**
 * @Component: Nav Group
 * @Description: This is in page navigation component to create
 * from Bootstrap.
*/
@Component({
  selector: 'citi-nav-group',
  templateUrl: './nav-group.component.html',
  styleUrls: ['./nav-group.component.css']
})
export class NavGroupComponent implements OnInit {

  /**
    * @Property: idStr 
    * @Description: This defines the id of in page navigation
    * @Type: string 
   */
  @Input() idStr:string;
   /**
    * @Property: activeTab 
    * @Description: This defines the active panel of in page navigation
    * @Type: string 
   */ 
  @Input() activeTab;
  private tabLength:Number;

  // Event emitter
  @Output() tabChanged: EventEmitter<any> = new EventEmitter();

  private tabs: NavPanelComponent[] = [];
  public defaultTab: number;

  addTab(tab:NavPanelComponent) {
    this.tabs.push(tab);
    
    return this.tabs.length;
  }
  countTabs() {
    return this.tabs.length;
  }

  constructor() {}
  
  ngOnInit() {
    this.tabLength = this.tabs.length
    this.defaultTab = this.activeTab ? Number.parseInt(this.activeTab,10) : 1;
    this.selectTab(this.tabs[this.defaultTab - 1], null);
    console.log("length",this.tabs.length);
  }
  
  selectTab(tab: NavPanelComponent, event) {
    if (event) {
      event.preventDefault();
    }
    this.tabs.forEach((tab) => {
      tab.isExpanded = false;
    });
    tab.isExpanded = true;

    // Event emitter
    this.tabChanged.emit(tab.tabIndex);
  }

  public currentTab() {
    this.tabs.forEach((tab) => {
      if (tab.isExpanded) {
        return tab.tabIndex;
      }
    });
    return -1;
  }

}
