// citi/forms/input-group/radio/radio.component.ts
import { Component, ViewChild, ElementRef, Input, OnInit } from '@angular/core';
import { InputService } from "../service/service.service";

/**
 * @Component: Radio Button
 * @Description: This component generates a radio button
 */
@Component({
    selector: 'citi-radio',
    template: require('./radio.component.html'),
    styles: [require("./radio.component.scss")]
})

export class RadioComponent {

    // A reference to the radio box
    @ViewChild("radio") radio: ElementRef;

    // Whether or not this radio button is a content addition
    isContentAddition: boolean = false;

    /**
     * @Property: contentAddition
     * @Description: Whethere the radio button is a content addition or not.
     * @Type: String
     */
    @Input()
    set contentAddition( any ){
        this.isContentAddition = true;
    }

    /**
     * @Property: id
     * @Description: The id for the radio button
     * @Type: String
     */
    @Input() id: string;

    /**
     * @Property: name
     * @Description: The name for the radio button
     * @Type: String
     */
    @Input() name: string;

    /**
     * @Property: value
     * @Description: The value for the radio button
     * @Type: String
     */
    @Input() value: string;
    
    /**
     * @Property: label
     * @Description: The label for the radio button
     * @Type: String
     */
    @Input() label: string;

    /**
     * @Property: disabled
     * @Description: Whether the radio button is disabled or not
     * @Type: String
     */
    @Input() disabled: boolean;

    /**
     * @Property: checked
     * @Description: Whether the radio button is checked or not
     * @Type: any
     */
    @Input() checked: any;
}
