// citi/jamp/jamp.component.ts
import { Component, OnInit, Input, HostBinding } from '@angular/core';

/**
 * @Component: JAMP
 * @Description: This component generates a JAMP (Just A Moment Please) element that can be shown during loading
 */

@Component({
  selector: 'citi-jamp',
  template: require('./jamp.component.html'),
  styles: [ require('./jamp.component.css') ]
})
export class JampComponent implements OnInit {
  // Set classes on host <citi-jamp> tag
  @HostBinding('class.hidden') isHidden = true;
  @HostBinding('class.jamp-page-level') isPageLevel = false;

  /**
   * @Property: useGif
   * @Description: Whether to show a GIF instead of CSS animation
   * @Type: Boolean
   */
  @Input() useGif: any;

  /**
   * @Property: visible
   * @Description: Whether JAMP is visible
   * @Type: Boolean
   */
  @Input() visible: any;

  /**
   * @Property: pagelevel
   * @Description: Whether JAMP fills the entire window instead of just its immediate container
   * @Type: Boolean
   */
  @Input() pagelevel: any;

  private jampCssClass: string = "";
  private loadingClass: string = "";

  constructor() {}

  ngOnInit() {
    if (!this.useGif) {
      this.jampCssClass = "jamp-css";
      this.loadingClass = "loading";
    }
    if (this.visible) {
      this.isHidden = false;
    }
    if (this.pagelevel) {
      this.isPageLevel = true;
    }
  }

  /**
    * @Function: show
    * @Description: Makes the JAMP visible.
    * @Return: Void
    */
  show() {
    this.isHidden = false;
  }

  /**
    * @Function: show
    * @Description: Makes the JAMP hidden.
    * @Return: Void
    */
  hide() {
    this.isHidden = true;
  }

}
