// ddl/card-module/figure/figure.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Card figure
 * @Section: Modules
 * @Description: This component creates a single figure for feature lists
 */
var FigureComponent = (function () {
    function FigureComponent() {
    }
    FigureComponent.prototype.ngOnInit = function () {
        this.footerClass = (this.footer ? "footer" : "");
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], FigureComponent.prototype, "label", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], FigureComponent.prototype, "footer", void 0);
    FigureComponent = __decorate([
        core_1.Component({
            selector: 'citi-card-figure',
            template: require('./figure.component.html'),
            //styleUrls: ['./figure.component.css']
            styles: [require('./figure.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], FigureComponent);
    return FigureComponent;
}());
exports.FigureComponent = FigureComponent;
