// ddl/branding/utils.class.js
"use strict";
var Utils = (function () {
    function Utils() {
        // Determins if a string is valid
        this.isValidString = function (string) {
            if (string
                && typeof string === "string"
                && string.length > 0) {
                return true;
            }
            else {
                return false;
            }
        };
        // This function determines if there is a valid social (facebook etc)
        // settings object in the social object it is passed
        this.hasValidSocialObject = function (social) {
            // Get the keys of the social object
            // This is used to loop over the object
            var socialKeys = Object.keys(social);
            // For each social object
            for (var i = 0, j = socialKeys.length; i < j; i++) {
                // Get the social object
                var socialObject = social[socialKeys[i]];
                // If it has a type of object 
                // (this is a sanity check against the show: true/false value)
                if (typeof socialObject === "object") {
                    // If it should be shown
                    if (this.isValidSocialObject(socialObject)) {
                        // Stop and return true since we found a valid object
                        return true;
                    }
                }
            }
        };
        // This function checks to see if a social object should be displayed
        this.isValidSocialObject = function (social) {
            // And if it should be shown
            if (social.show
                && this.isValidString(social.header)
                && this.isValidString(social.copy)) {
                // Stop and return true since we found a valid object
                return true;
            }
        };
    }
    return Utils;
}());
exports.Utils = Utils;
