// ddl/progress/step.js
"use strict";
(function (StepState) {
    StepState[StepState["Active"] = 0] = "Active";
    StepState[StepState["Finished"] = 1] = "Finished";
    StepState[StepState["Normal"] = 2] = "Normal";
})(exports.StepState || (exports.StepState = {}));
var StepState = exports.StepState;
var Step = (function () {
    function Step(text, state) {
        if (state === void 0) { state = StepState.Normal; }
        this.state = StepState.Normal;
        this.setText(text);
    }
    Step.prototype.getText = function () {
        return this.text;
    };
    Step.prototype.getState = function () {
        return this.state;
    };
    Step.prototype.setText = function (text) {
        this.text = text;
    };
    Step.prototype.setState = function (state) {
        if (state === void 0) { state = StepState.Normal; }
        this.state = state;
    };
    Step.prototype.setActive = function () {
        this.state = StepState.Active;
    };
    Step.prototype.setFinished = function () {
        this.state = StepState.Finished;
    };
    Step.prototype.setNormal = function () {
        this.state = StepState.Normal;
    };
    Step.prototype.isActive = function () {
        return this.state === StepState.Active;
    };
    Step.prototype.isFinished = function () {
        return this.state === StepState.Finished;
    };
    return Step;
}());
exports.Step = Step;
