// citi/in-page-navigation/nav-panel/nav-panel.component.ts
import { Component, Input, OnInit, HostBinding } from '@angular/core';
import { NavGroupComponent } from '../nav-group/nav-group.component';

/**
 * @Component: Nav Panel
 * @Description: This is the nav panel component and is used to create panel in the nav group
 * from Bootstrap.
*/
@Component({
  selector: 'citi-nav-panel',
  templateUrl: './nav-panel.component.html',
  styleUrls: ['./nav-panel.component.css']
})
export class NavPanelComponent implements OnInit {
 // Set classes on host <citi-tab> tag
  @HostBinding('class.tab-pane') classTabPane = true;
  @HostBinding('class.level1') classLevel1 = true;
  @HostBinding('class.parent') classParent = true;
  @HostBinding('class.active') isExpanded = false;

  /**
    * @Property: title 
    * @Description: This defines the title of the in page navigation
    * @Type: string 
   */
  @Input('title') tabTitle;
  private tabgroup: NavGroupComponent;
  private lastTab: Boolean;
  public tabIndex: number;

  constructor(private parent: NavGroupComponent) {
    this.tabgroup = parent;
    this.tabIndex = this.tabgroup.addTab(this);
  }

  ngOnInit() {
    var totalTabs: number = this.tabgroup.countTabs();
    this.lastTab = this.tabIndex === totalTabs;
    this.isExpanded = this.tabIndex === this.tabgroup.defaultTab;
  }

  selectTab(event) {
    this.tabgroup.selectTab(this, event);
  }
}
