// ddl/branding/footer/disclaimer/disclaimer.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Footer Disclaimer
 * @Section: Navigation
 * @Description: This component generates the disclaimer section in the footer.
 */
var DisclaimerBrandingComponent = (function () {
    function DisclaimerBrandingComponent() {
        // This function takes the BrandingConfig object and translates it to config DOM
        // This step is not necessary, but keeps the DOM markup semantic and un-bloated
        // and keeps most of the logic in the controller instead of the view
        this.setDisclaimerConfig = function (_config) {
            // Set up config object to temporarily hold our calculated values
            var disclaimerConfig = {};
            // If there is a valid config and a footer object is defined
            if (_config && _config.footer) {
                // If the butler bar should be shown
                if (_config.footer.disclaimer
                    && _config.footer.disclaimer.show) {
                    // If there is a footer disclaimer sections array and it has at least one section
                    if (_config.footer.disclaimer.text) {
                        // Set the disclaimer to show
                        disclaimerConfig.showDisclaimer = true;
                        // Provide the sections to the config object
                        disclaimerConfig.text = _config.footer.disclaimer.text;
                    }
                }
                // Set the config for use by the dom (DOM will not show until this is set);
                this.config = disclaimerConfig;
            }
        };
    }
    Object.defineProperty(DisclaimerBrandingComponent.prototype, "brandingConfig", {
        /**
         * @Property: brandingConfig
         * @Description: This is a reference to the branding config JSON object retrieved from the branding API.
         * @Type: String
         */
        set: function (_config) {
            if (_config) {
                this.setDisclaimerConfig(_config);
            }
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], DisclaimerBrandingComponent.prototype, "brandingConfig", null);
    DisclaimerBrandingComponent = __decorate([
        core_1.Component({
            selector: 'citi-footer-disclaimer',
            template: require('./disclaimer.component.html'),
            styles: [require("./disclaimer.component.scss")]
        }), 
        __metadata('design:paramtypes', [])
    ], DisclaimerBrandingComponent);
    return DisclaimerBrandingComponent;
}());
exports.DisclaimerBrandingComponent = DisclaimerBrandingComponent;
