// ddl/forms/input-group/input.service.js
"use strict";
var core_1 = require('@angular/core');
var InputService = (function () {
    function InputService() {
        this.errors = [];
        this.inputs = [];
        this.addError = function (error) {
            if (this.errors.indexOf(error) === -1) {
                this.errors.push(error);
                this.error = true;
            }
            ;
        };
        this.removeError = function (error) {
            if (this.errors.indexOf(error) !== -1) {
                this.errors.splice(this.errors.indexOf(error), 1);
                this.errors.length === 0 ? this.error = false : null;
            }
        };
        this.addInput = function (input) {
            this.inputs.push(input);
        };
        this.validate = function () {
            this.validateNow.emit();
        };
        this.errorAdded = new core_1.EventEmitter();
        this.errorRemoved = new core_1.EventEmitter();
        this.validateNow = new core_1.EventEmitter();
    }
    return InputService;
}());
exports.InputService = InputService;
