// citi/hero/hero.component.ts
import { Component, OnInit, Input, ViewChild, ElementRef } from '@angular/core';

@Component({
  selector: 'citi-hero',
  template: require('./hero.component.html'),
  styles: [ require('./hero.component.css') ]
})
export class HeroComponent implements OnInit {

  private type: 'a-1'| 'a-2' | 'a-3' | 'a-4' = 'a-1';
  private promoImage: string = '';
  private footer: string;
  private eyebrow: string;
  private legend: string;
  private shows: boolean = true;
  @ViewChild('content') content: ElementRef;
  @Input()
  set promoImageSrc ( promoImage: string ){
    this.promoImage = promoImage;
  }
  @Input()
  set title ( title: string ){
    this.legend = title;
  }
  @Input()
  set heroType ( type: any ){
    this.type = type;
  }
  @Input()
  set footerContent ( footerContent: string ){
    this.footer = footerContent;
  }
  @Input()
  set eyebrowText ( eyebrowText: string ){
    this.eyebrow = eyebrowText;
  }

  constructor() { }

  ngOnInit() {
  }
  ngAfterViewInit() {
    debugger;
  }
}
