// ddl/in-page-navigation/nav-panel/nav-panel.component.js
"use strict";
var core_1 = require('@angular/core');
var nav_group_component_1 = require('../nav-group/nav-group.component');
/**
 * @Component: Nav Panel
 * @Section: Navigation
 * @Description: This is the nav panel component and is used to create panel in the nav group
 * from Bootstrap.
*/
var NavPanelComponent = (function () {
    function NavPanelComponent(parent) {
        this.parent = parent;
        // Set classes on host <citi-tab> tag
        this.classTabPane = true;
        this.classLevel1 = true;
        this.classParent = true;
        this.isExpanded = false;
        this.tabgroup = parent;
        this.tabIndex = this.tabgroup.addTab(this);
    }
    NavPanelComponent.prototype.ngOnInit = function () {
        var totalTabs = this.tabgroup.countTabs();
        this.lastTab = this.tabIndex === totalTabs;
        this.isExpanded = this.tabIndex === this.tabgroup.defaultTab;
    };
    NavPanelComponent.prototype.selectTab = function (event) {
        this.tabgroup.selectTab(this, event);
    };
    __decorate([
        core_1.HostBinding('class.tab-pane'), 
        __metadata('design:type', Object)
    ], NavPanelComponent.prototype, "classTabPane", void 0);
    __decorate([
        core_1.HostBinding('class.level1'), 
        __metadata('design:type', Object)
    ], NavPanelComponent.prototype, "classLevel1", void 0);
    __decorate([
        core_1.HostBinding('class.parent'), 
        __metadata('design:type', Object)
    ], NavPanelComponent.prototype, "classParent", void 0);
    __decorate([
        core_1.HostBinding('class.active'), 
        __metadata('design:type', Object)
    ], NavPanelComponent.prototype, "isExpanded", void 0);
    __decorate([
        core_1.Input('title'), 
        __metadata('design:type', Object)
    ], NavPanelComponent.prototype, "tabTitle", void 0);
    NavPanelComponent = __decorate([
        core_1.Component({
            selector: 'citi-nav-panel',
            template: require('./nav-panel.component.html'),
            //styleUrls: ['./nav-panel.component.css']
            styles: [require('./nav-panel.component.css')]
        }), 
        __metadata('design:paramtypes', [nav_group_component_1.NavGroupComponent])
    ], NavPanelComponent);
    return NavPanelComponent;
}());
exports.NavPanelComponent = NavPanelComponent;
