// ddl/image/image.component.js
"use strict";
var core_1 = require('@angular/core');
var ImageComponent = (function () {
    function ImageComponent() {
        this.imageLeftClass = 'pull-out-quote pull-quote-content-p-1';
        this.imageRightClass = 'pull-out-quote pull-quote-content-p-2';
        this.imageCenterClass = 'pull-out-quote-3 pull-out-quote-p-3';
        this.imageClass = this.imageLeftClass;
        this.imageType = 'left';
        this.quoteBool = true;
        this.imageUrl = 'images/pull-out-quote-v2.png';
        this.quoteText = 'Quote text';
    }
    Object.defineProperty(ImageComponent.prototype, "type", {
        set: function (type) {
            this.imageType = type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageComponent.prototype, "quote", {
        set: function (value) {
            if (typeof value == "boolean" || value == "true" || value == "false") {
                if (typeof value == 'string') {
                    this.quoteBool = (value == "true");
                }
                else if (typeof value == 'boolean') {
                    this.quoteBool = value;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    ;
    ImageComponent.prototype.ngOnInit = function () {
        switch (this.imageType) {
            case 'left':
                this.imageClass = this.imageLeftClass;
                break;
            case 'right':
                this.imageClass = this.imageRightClass;
                break;
            case 'center':
                this.imageClass = this.imageCenterClass;
                break;
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], ImageComponent.prototype, "type", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ImageComponent.prototype, "imageUrl", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ImageComponent.prototype, "quoteText", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], ImageComponent.prototype, "quote", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ImageComponent.prototype, "altText", void 0);
    ImageComponent = __decorate([
        core_1.Component({
            selector: 'citi-image',
            template: require('./image.component.html')
        }), 
        __metadata('design:paramtypes', [])
    ], ImageComponent);
    return ImageComponent;
}());
exports.ImageComponent = ImageComponent;
