// ddl/alert/alert.component.js
"use strict";
var core_1 = require('@angular/core');
var AlertComponent = (function () {
    function AlertComponent() {
        this.alertClass = 'alert';
        this.criticalClass = 'critical';
        this.alertBoxClass = 'alert-box';
        this.alertType = 'alert';
        this.notice = false;
        this.alertText = 'Alert!';
    }
    Object.defineProperty(AlertComponent.prototype, "type", {
        set: function (type) {
            this.alertType = type;
        },
        enumerable: true,
        configurable: true
    });
    AlertComponent.prototype.ngOnInit = function () {
        switch (this.alertType) {
            case 'alert':
                this.cssClass = this.alertClass;
                break;
            case 'critical':
                this.cssClass = this.alertClass + ' ' + this.criticalClass;
                break;
            case 'alertBox':
                this.cssClass = this.alertBoxClass;
                break;
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], AlertComponent.prototype, "type", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AlertComponent.prototype, "notice", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AlertComponent.prototype, "noticeText", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AlertComponent.prototype, "alertText", void 0);
    AlertComponent = __decorate([
        core_1.Component({
            selector: 'citi-alert',
            template: require('./alert.component.html')
        }), 
        __metadata('design:paramtypes', [])
    ], AlertComponent);
    return AlertComponent;
}());
exports.AlertComponent = AlertComponent;
