// citi/forms/input/input.component.classes.ts
export class InputSettings {
    type: 'standard' | 'tooltip' | 'finAddOn' | 'password' | 'simpleDate' | 'email' | 'url' | 'phoneNumber' | 'phoneCountry' | 'phoneArea' | 'readOnly' | 'error' | 'ssn' = 'standard';
    size: 'large' | 'medium' | 'small' | 'xsmall' | 'none' = 'large';
    mask: 'none' | 'phone' | 'date' | 'zipcode' | 'dollar' | 'email' = 'none';
    pipe: 'none' | 'ssn' = 'none';
}

export class InputCss {
    public static get Large(): string {return 'col-md-8 col-sm-12';}
    public static get Medium(): string {return 'col-md-6 col-sm-9 col-xs-12';}
    public static get Small(): string {return 'col-md-4 col-sm-6 col-xs-12';}
    public static get Success(): string {return 'validation-input-success';}
    public static get InputGroup(): string {return 'input-group';}
    public static get InputGroupFocus(): string {return 'focus-group';}
    public static get TextRight(): string {return 'text-right';}
}

export class InputSize {
    public static get Large(): string {return 'large';}
    public static get Medium(): string {return 'medium';}
    public static get Small(): string {return 'small';}
    public static get XSmall(): string {return 'xsmall';}
}

export class InputType {
    public static get Standard(): string {return 'standard';}
    public static get Tooltip(): string {return 'tooltip';}
    public static get FinAddOn(): string {return 'finAddOn';}
    public static get Password(): string {return 'password';}
    public static get SimpleDate(): string {return 'simpleDate';}
    public static get Email(): string {return 'email';}
    public static get Url(): string {return 'url';}
    public static get PhoneNumber(): string {return 'phoneNumber';}
    public static get ReadOnly(): string {return 'readOnly';}
    public static get Error(): string {return 'error';}
}

export class TooltipInput {
    public static get InputCss(): string {return '';}
    public static get PlacementTop(): string {return 'auto top';}
    public static get PlacementBot(): string {return 'auto bottom';}
    public static get InputGroupCss(): string {return 'tooltip-input-group';}
}

export class FinAddOnInput {
    public static get Step(): number {return 1;}
    public static get Min(): number {return 0;}
    public static get TypeDollar(): string {return '$';}
    public static get TypeRUB(): string {return 'RUB';}
    public static get InputGroupCss(): string {return 'input-group focus-group';}
}

export class PasswordInput {
    public static get InputCss(): string {return 'text-input-container';}
    public static get MaxLength(): number {return 10;}
    public static get Type(): string {return 'password';}
    public static get LargeRules(): string {return 'no-spaces min-max-8-10 has-letter has-number';}
    public static get MediumRules(): string {return 'hasspecialcharacter no-consecutive-3 no-spaces min-max-8-10 has-letter has-number';}
    public static get SmallRules(): string {return 'min-max-6-25';}
    public static get XSmallRules(): string {return 'no-consecutive no-special-characters has-letter has-number';}
    public static get PlacementTop(): string {return 'top auto';}
    public static get SuccessMsg(): string {return 'Your password meets all tested requirements';}
    public static get FailMsg(): string {return 'Your password doesn&#39;t meet the requirements below';}
}

export class SimpleDateInput {
    public static get RowCss(): string {return 'cbolui-input-box';}
    public static get Type(): string {return 'text';}
    public static get InputCss(): string {return 'dateinput-simple form-control formatted';}
    public static get Label1(): string {return 'Date (dd/mm/yyyy)';}
    public static get Placeholder1(): string {return 'Date dd/mm/yyyy';}
    public static get Format1(): string {return 'dd/mm/yyyy';}
    public static get AddFormat1(): string {return 'dd/mm/y|dd-mm-y|dd-mm-yy';}
    public static get Label2(): string {return 'Date (mm/dd/yy)';}
    public static get Placeholder2(): string {return 'Date mm/dd/yy';}
    public static get Format2(): string {return 'mm/dd/yy';}
}

export class EmailInput {
    public static get Type(): string {return 'email';}
    public static get InputCss(): string {return 'formatted';}
    public static get Label(): string {return 'Email Address (name@email.com)';}
    public static get Placeholder(): string {return 'Email Address';}
    public static get Format(): string {return 'name@email.com';}
}

export class UrlInput {
    public static get Type(): string {return 'url';}
    public static get InputCss(): string {return 'formatted';}
    public static get Label(): string {return 'URL (http://www.google.com)';}
    public static get Placeholder(): string {return 'URL';}
    public static get Format(): string {return 'http://www.google.com';}
}

export class PhoneInput {
    public static get Type(): string {return 'tel';}
    public static get InputCss(): string {return 'formatted';}
    public static get Label(): string {return 'Telephone Number';}
    public static get Placeholder(): string {return 'Telephone Number';}
    public static get Format(): string {return '416 673 5275';}
    public static get MaxLength(): number {return 16;}
    public static get FormatEnforcement(): string {return '[\d ]';}
}

export class ReadOnlyInput {
    public static get InputCss(): string {return 'read-only';}
}

export class ErrorInput {
    public static get InputCss(): string {return 'validation-input-danger';}
}
