// citi/forms/input-group/checkbox/checkbox.component.ts
import { Component, Input } from '@angular/core';

/**
 * @Component: Checkbox
 * @Description: This component generates the checkbox
 */
@Component({
    selector: 'citi-checkbox',
    template: require('./checkbox.component.html'),
    styles: [require("./checkbox.component.scss")]
})

export class CheckboxComponent {

    // Whether or not this radio button is a content addition
    isContentAddition: Boolean = false;

    /**
     * @Property: contentAddition
     * @Description: Where the checkbox is a content addition or not.
     * @Type: String
     */
    @Input() 
    set contentAddition( any ){
        this.isContentAddition = true;
    }

    /**
     * @Property: id
     * @Description: The id for the checkbox
     * @Type: String
     */
    @Input() idStr: string;

    /**
     * @Property: name
     * @Description: The name for the checkbox
     * @Type: String
     */
    @Input() name: string;

    /**
     * @Property: value
     * @Description: The value for the checkbox
     * @Type: String
     */
    @Input() value: string;
    
    /**
     * @Property: label
     * @Description: The label for the checkbox
     * @Type: String
     */
    @Input() label: string;

    /**
     * @Property: disabled
     * @Description: Whether the checkbox is disabled or not
     * @Type: String
     */
    @Input() disabled: boolean;

    /**
     * @Property: checked
     * @Description: Whether the checkbox is checked or not
     * @Type: String
     */
    @Input() checked: boolean;
}
