// ddl/card-module/highlight/highlight.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Card highlights
 * @Section: Modules
 * @Description: This component creates a single column of card highlights for the C2.X DDL modules
 */
var HighlightComponent = (function () {
    function HighlightComponent() {
        // Does this module have a "callout" to display
        this.hasCallout = true;
    }
    HighlightComponent.prototype.ngOnInit = function () {
        this.hasCallout = this.callout.nativeElement.childNodes.length !== 0;
    };
    __decorate([
        core_1.ViewChild('callout'), 
        __metadata('design:type', core_1.ElementRef)
    ], HighlightComponent.prototype, "callout", void 0);
    HighlightComponent = __decorate([
        core_1.Component({
            selector: 'citi-card-highlight',
            template: require('./highlight.component.html'),
            //styleUrls: ['./highlight.component.css']
            styles: [require('./highlight.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], HighlightComponent);
    return HighlightComponent;
}());
exports.HighlightComponent = HighlightComponent;
