// ddl/forms/table/expand/expand.component.js
"use strict";
var core_1 = require('@angular/core');
var ExpandComponent = (function () {
    function ExpandComponent() {
        this._isExpanded = false;
    }
    Object.defineProperty(ExpandComponent.prototype, "textOpen", {
        get: function () {
            return this._textOpen;
        },
        set: function (value) {
            this._textOpen = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandComponent.prototype, "textCollapse", {
        get: function () {
            return this._textCollapse;
        },
        set: function (value) {
            this._textCollapse = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpandComponent.prototype, "isExpanded", {
        get: function () {
            return this._isExpanded;
        },
        set: function (value) {
            this._isExpanded = value;
        },
        enumerable: true,
        configurable: true
    });
    ExpandComponent.prototype.onClickBtn = function (btn) {
        if (btn.tagName === 'SPAN' && (' ' + btn.className + ' ').indexOf(' btn-link ') > -1) {
            this.isExpanded = !this.isExpanded;
        }
    };
    ExpandComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ExpandComponent.prototype, "textOpen", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ExpandComponent.prototype, "textCollapse", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], ExpandComponent.prototype, "isExpanded", null);
    __decorate([
        core_1.HostListener('click', ['$event.target']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object]), 
        __metadata('design:returntype', void 0)
    ], ExpandComponent.prototype, "onClickBtn", null);
    ExpandComponent = __decorate([
        core_1.Component({
            selector: 'citi-expand',
            template: require('./expand.component.html'),
            //styleUrls: ['./expand.component.css']
            styles: [require('./expand.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], ExpandComponent);
    return ExpandComponent;
}());
exports.ExpandComponent = ExpandComponent;
