// ddl/copy/blocks/blocks.js
"use strict";
var copy_component_1 = require("./copy/copy.component");
exports.CopyComponent = copy_component_1.CopyComponent;
var list_component_1 = require("./list/list.component");
exports.ListComponent = list_component_1.ListComponent;
var important_copy_component_1 = require("./important-copy/important-copy.component");
exports.ImportantCopyComponent = important_copy_component_1.ImportantCopyComponent;
var accordion_1 = require("./accordion/accordion");
exports.AccordionComponent = accordion_1.AccordionComponent;
exports.AccordionSectionComponent = accordion_1.AccordionSectionComponent;
var sub_content_1 = require("./sub-content/sub-content");
exports.SubContentComponent = sub_content_1.SubContentComponent;
exports.SubContentSectionComponent = sub_content_1.SubContentSectionComponent;
var two_column_list_1 = require("./two-column-list/two-column-list");
exports.TwoColumnListComponent = two_column_list_1.TwoColumnListComponent;
exports.TwoColumnListSectionComponent = two_column_list_1.TwoColumnListSectionComponent;
var two_column_content_1 = require("./two-column-content/two-column-content");
exports.TwoColumnContentComponent = two_column_content_1.TwoColumnContentComponent;
exports.TwoColumnTitleSectionComponent = two_column_content_1.TwoColumnTitleSectionComponent;
var two_column_title_grid_1 = require("./two-column-title-grid/two-column-title-grid");
exports.TwoColumnTitleGridComponent = two_column_title_grid_1.TwoColumnTitleGridComponent;
exports.TitleGridSectionComponent = two_column_title_grid_1.TitleGridSectionComponent;
