// citi/utility/services/masking.service.ts
import { Injectable } from '@angular/core';

import { MaskTypes, createSSNPipe } from '../../utility/utility.classes';
import emailMask from 'text-mask-addons/dist/emailMask.js'

@Injectable()
export class MaskingService {
    getMask: Function = function(type){
        switch (type) {
            case "phone":
                return MaskTypes.phoneMask;
            case "date":
                return MaskTypes.dateMask;
            case "zipcode":
                return MaskTypes.zipcodeMask;
            case "dollar":
                return MaskTypes.dollarMask;
            case "email":
                return emailMask;
            case "ssn":
                return MaskTypes.ssnMask;
        }
    }

    getPhoneMask(): any[] {
        return MaskTypes.phoneMask;
    }
    getDateMask(): any[] {
        return MaskTypes.dateMask;
    }
    getZipcodeMask(): any[] {
        return MaskTypes.zipcodeMask;
    }
    getDollarMask(): any[] {
        return MaskTypes.dollarMask;
    }
    getEmailMask(): any[] {
        return emailMask;
    }
    getSSNPipe(): any {
        return createSSNPipe;
    }
}
