// ddl/navigations/left-navigation/left-navigation-child/left-nav-child.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Left Nav Child
 * @Section: Navigation
 * @Description: This is first level child component inside the parent component.
*/
var ChildComponent = (function () {
    function ChildComponent() {
        this.isExpanded = false;
        // Custom Attributes
        this.hasChildren = false;
    }
    // When span clicked
    ChildComponent.prototype.isClicked = function (event) {
        this.isExpanded = !this.isExpanded;
    };
    /**
    * @Function: logClick
    * @Description: This function stops propagating the event in the DOM hierarchy.
    * @Param: (String) - event - Event object.
    * @Param: (String) - index - Index of clicked nav item.
    * @Return: -
  */
    ChildComponent.prototype.logClick = function (event, index) {
        event.stopPropagation();
    };
    ChildComponent.prototype.ngOnInit = function () { };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], ChildComponent.prototype, "hasChildren", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ChildComponent.prototype, "title", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ChildComponent.prototype, "href", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Array)
    ], ChildComponent.prototype, "subMenuList", void 0);
    ChildComponent = __decorate([
        core_1.Component({
            selector: '.citi-left-nav-child',
            template: require('./left-nav-child.component.html'),
            styles: [require('./left-nav-child.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], ChildComponent);
    return ChildComponent;
}());
exports.ChildComponent = ChildComponent;
