// ddl/tabs/tab-group/tab-group.component.js
"use strict";
var core_1 = require('@angular/core');
/**
 * @Component: Tab-group
 * @Description: A wrapper for displaying individual tab panels
 */
var TabGroupComponent = (function () {
    function TabGroupComponent() {
        /**
          * @Event: tabChanged
          * @Description: This event that is fired when a tab activates; returns the index of the active tab (first tab is 1)
          * @Return: Number
          */
        this.tabChanged = new core_1.EventEmitter();
        this.tabs = [];
    }
    TabGroupComponent.prototype.addTab = function (tab) {
        this.tabs.push(tab);
        return this.tabs.length;
    };
    TabGroupComponent.prototype.countTabs = function () {
        return this.tabs.length;
    };
    TabGroupComponent.prototype.ngOnInit = function () {
        this.defaultTab = this.activeTab ? Number.parseInt(this.activeTab, 10) : 1;
        this.selectTab(this.tabs[this.defaultTab - 1], null);
    };
    TabGroupComponent.prototype.selectTab = function (tab, event) {
        if (event) {
            event.preventDefault();
        }
        this.tabs.forEach(function (tab) {
            tab.isExpanded = false;
        });
        tab.isExpanded = true;
        // Event emitter
        this.tabChanged.emit(tab.tabIndex);
    };
    TabGroupComponent.prototype.currentTab = function () {
        this.tabs.forEach(function (tab) {
            if (tab.isExpanded) {
                return tab.tabIndex;
            }
        });
        return -1;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], TabGroupComponent.prototype, "id", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], TabGroupComponent.prototype, "activeTab", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], TabGroupComponent.prototype, "tabChanged", void 0);
    TabGroupComponent = __decorate([
        core_1.Component({
            selector: 'citi-tab-group',
            template: require('./tab-group.component.html')
        }), 
        __metadata('design:paramtypes', [])
    ], TabGroupComponent);
    return TabGroupComponent;
}());
exports.TabGroupComponent = TabGroupComponent;
