// ddl/copy/blocks/accordion/accordion-section/accordion-section.component.js
"use strict";
var core_1 = require('@angular/core');
var AccordionSectionComponent = (function () {
    // Create a reference to this element
    function AccordionSectionComponent(element) {
        this.element = element;
    }
    ;
    AccordionSectionComponent.prototype.ngOnInit = function () {
        // Get all of the sections in this accordion
        var sections = this.element.nativeElement.parentElement.getElementsByTagName("citi-accordion-section");
        // Convert the HTMLCollection into an array so we can use indexOf
        sections = [].slice.call(sections);
        // Set the index of this section
        this.index = sections.indexOf(this.element.nativeElement);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AccordionSectionComponent.prototype, "title", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AccordionSectionComponent.prototype, "description", void 0);
    AccordionSectionComponent = __decorate([
        core_1.Component({
            selector: 'citi-accordion-section',
            template: require('./accordion-section.component.html')
        }), 
        __metadata('design:paramtypes', [core_1.ElementRef])
    ], AccordionSectionComponent);
    return AccordionSectionComponent;
}());
exports.AccordionSectionComponent = AccordionSectionComponent;
