// ddl/branding/footer/footer.component.js
"use strict";
var core_1 = require('@angular/core');
var branding_service_1 = require("../services/branding.service");
var FooterComponent = (function () {
    // Constructor
    function FooterComponent(brandingService) {
        this.brandingService = brandingService;
    }
    ;
    FooterComponent.prototype.ngOnInit = function () {
        var _this = this;
        // Get the endpoint url
        var endpointUrl = this.brandingService.getCurrentUrl() || this.url;
        // If we found an endpoint url
        if (endpointUrl) {
            // Get data from the branding service
            this.brandingService.getData(endpointUrl).subscribe(function (data) {
                _this.brandingConfig = data;
            });
        }
    };
    Object.defineProperty(FooterComponent.prototype, "endpoint", {
        set: function (url) {
            this.url = url;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], FooterComponent.prototype, "endpoint", null);
    FooterComponent = __decorate([
        core_1.Component({
            selector: 'citi-footer',
            template: require('./footer.component.html')
        }), 
        __metadata('design:paramtypes', [branding_service_1.BrandingService])
    ], FooterComponent);
    return FooterComponent;
}());
exports.FooterComponent = FooterComponent;
