// ddl/hero/hero.component.js
"use strict";
var core_1 = require('@angular/core');
var HeroComponent = (function () {
    function HeroComponent() {
        this.type = 'a-1';
        this.promoImage = '';
        this.shows = true;
    }
    Object.defineProperty(HeroComponent.prototype, "promoImageSrc", {
        set: function (promoImage) {
            this.promoImage = promoImage;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeroComponent.prototype, "title", {
        set: function (title) {
            this.legend = title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeroComponent.prototype, "heroType", {
        set: function (type) {
            this.type = type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeroComponent.prototype, "footerContent", {
        set: function (footerContent) {
            this.footer = footerContent;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeroComponent.prototype, "eyebrowText", {
        set: function (eyebrowText) {
            this.eyebrow = eyebrowText;
        },
        enumerable: true,
        configurable: true
    });
    HeroComponent.prototype.ngOnInit = function () {
    };
    HeroComponent.prototype.ngAfterViewInit = function () {
        debugger;
    };
    __decorate([
        core_1.ViewChild('content'), 
        __metadata('design:type', core_1.ElementRef)
    ], HeroComponent.prototype, "content", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], HeroComponent.prototype, "promoImageSrc", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], HeroComponent.prototype, "title", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], HeroComponent.prototype, "heroType", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], HeroComponent.prototype, "footerContent", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], HeroComponent.prototype, "eyebrowText", null);
    HeroComponent = __decorate([
        core_1.Component({
            selector: 'citi-hero',
            template: require('./hero.component.html'),
            styles: [require('./hero.component.css')]
        }), 
        __metadata('design:paramtypes', [])
    ], HeroComponent);
    return HeroComponent;
}());
exports.HeroComponent = HeroComponent;
