// angular2-nexus-uiux/core-app.module.js
/**
 * @module CoreAppModule
 * @main CoreAppModule
 * @preferred
 **/
/** CoreAppModule **/
"use strict";
var hmr_1 = require('@angularclass/hmr');
var CoreAppModule = (function () {
    function CoreAppModule(appRef, appStateService) {
        this.appRef = appRef;
        this.appStateService = appStateService;
    }
    CoreAppModule.prototype.hmrOnInit = function (store) {
        if (!store || !store.state)
            return;
        console.log('HMR store', JSON.stringify(store, null, 2));
        // set state
        this.appStateService._state = store.state;
        // set input values
        if ('restoreInputValues' in store) {
            var restoreInputValues = store.restoreInputValues;
            setTimeout(restoreInputValues);
        }
        this.appRef.tick();
        delete store.state;
        delete store.restoreInputValues;
    };
    CoreAppModule.prototype.hmrOnDestroy = function (store) {
        var cmpLocation = this.appRef.components.map(function (cmp) { return cmp.location.nativeElement; });
        // save state
        var state = this.appStateService._state;
        store.state = state;
        // recreate root elements
        store.disposeOldHosts = hmr_1.createNewHosts(cmpLocation);
        // save input values
        store.restoreInputValues = hmr_1.createInputTransfer();
        // remove styles
        hmr_1.removeNgStyles();
    };
    CoreAppModule.prototype.hmrAfterDestroy = function (store) {
        // display new elements
        store.disposeOldHosts();
        delete store.disposeOldHosts;
    };
    return CoreAppModule;
}());
exports.CoreAppModule = CoreAppModule;
