// ddl/tabs/tab/tab.component.js
"use strict";
var core_1 = require('@angular/core');
var tab_group_component_1 = require('../tab-group/tab-group.component');
/**
 * @Component: Tab
 * @Description: Individual tab panel to be wrapped in a citi-tab-group
 */
var TabComponent = (function () {
    function TabComponent(parent) {
        this.parent = parent;
        // Set classes on host <citi-tab> tag
        this.classTabPane = true;
        this.classLevel1 = true;
        this.classParent = true;
        this.isExpanded = false;
        this.tabgroup = parent;
        this.tabIndex = this.tabgroup.addTab(this);
    }
    TabComponent.prototype.ngOnInit = function () {
        var totalTabs = this.tabgroup.countTabs();
        this.lastTab = this.tabIndex === totalTabs;
        this.isExpanded = this.tabIndex === this.tabgroup.defaultTab;
    };
    TabComponent.prototype.selectTab = function (event) {
        this.tabgroup.selectTab(this, event);
    };
    __decorate([
        core_1.Input('title'), 
        __metadata('design:type', Object)
    ], TabComponent.prototype, "tabTitle", void 0);
    __decorate([
        core_1.HostBinding('class.tab-pane'), 
        __metadata('design:type', Object)
    ], TabComponent.prototype, "classTabPane", void 0);
    __decorate([
        core_1.HostBinding('class.level1'), 
        __metadata('design:type', Object)
    ], TabComponent.prototype, "classLevel1", void 0);
    __decorate([
        core_1.HostBinding('class.parent'), 
        __metadata('design:type', Object)
    ], TabComponent.prototype, "classParent", void 0);
    __decorate([
        core_1.HostBinding('class.is-expanded'), 
        __metadata('design:type', Object)
    ], TabComponent.prototype, "isExpanded", void 0);
    TabComponent = __decorate([
        core_1.Component({
            selector: 'citi-tab',
            template: require('./tab.component.html'),
            styles: [require('./tab.component.css')]
        }), 
        __metadata('design:paramtypes', [tab_group_component_1.TabGroupComponent])
    ], TabComponent);
    return TabComponent;
}());
exports.TabComponent = TabComponent;
