// citi/image/image.component.ts
import { Component, OnInit, Input } from '@angular/core';

@Component({
  selector: 'citi-image',
  template: require('./image.component.html')
})
export class ImageComponent implements OnInit {

  private imageLeftClass:string = 'pull-out-quote pull-quote-content-p-1';
  private imageRightClass:string = 'pull-out-quote pull-quote-content-p-2';
  private imageCenterClass:string = 'pull-out-quote-3 pull-out-quote-p-3';
  private imageClass:string = this.imageLeftClass;

  private imageType: 'left' | 'right' | 'center' = 'left';
  private quoteBool:Boolean = true;

  @Input()
  set type(type:any) {
    this.imageType = type;
  }

  @Input() imageUrl:string = 'images/pull-out-quote-v2.png';
  @Input() quoteText:string = 'Quote text';

  @Input() 
  set quote( value: any ){
    if (typeof value == "boolean" || value == "true" || value == "false"){
      if(typeof value == 'string') {        
        this.quoteBool = (value == "true");
      } else if(typeof value == 'boolean') {
        this.quoteBool = value;        
      }      
    }
  };

  @Input() altText:string;

  constructor() { }

  ngOnInit() {
    switch(this.imageType) {
      case 'left':
        this.imageClass = this.imageLeftClass;
        break;
      case 'right':
        this.imageClass = this.imageRightClass;
        break;
      case 'center':
        this.imageClass = this.imageCenterClass;
        break;
    }

  }

}
